[RFC 0166] Nix formatting, take two (#166)
NixOS/rfcs: rfcs/0166-nix-formatting.md

Reviewer: What is the [Hattie Effect Size](https://www.slideshare.net/richardcookau/john-hattie-effect-sizes-on-achievement) for any of our rules in readability?

ie:
What is the actual impact of 60 ~ 80 ~ 100 ~ :infinity: max len in readability?, None? Choose one, and move on…
What is the actual impact of vertical alignment? None? Remove for consistency…

This was what I could found about the subject.
https://arxiv.org/abs/2208.12141

What are our actual statics?
ie: How many files follows X length limit?

Reviewer 2: This is already being discussed in https://github.com/NixOS/rfcs/pull/166#discussion_r1395667716

Reviewer 2: I'll start a new thread for @quantenzitrone's [top-level comment](https://github.com/NixOS/rfcs/pull/166#pullrequestreview-1757403665) for it to be more manageable:

> Has sorting of attrs and lists been talked about? I think it would be very beneficial for a lot of nix code to have functionality to automatically sort lists or attribute sets. I also think this RFC could be a great place to establish a syntax for that. e.g.
> 
> ```nix
> {
>   foo = { # sort
>     # values here will be sorted
>   };
>   bar = { # sort rec
>     # everything will be sorted recursively (also nested attributes)
>   };
>   # the same for lists?
> }
> ```
> 
> (this syntax is not much thought about and probably not optimal)
> 
> Stuff to think about:
> 
> * What about case sensitivity, direction, human counting sorting?
> 
> * Maybe support sorting only a specific section of a list/attrs?

I think this would blow up the scope larger than it needs to be. The key to success of an RFC is to remove everything that's not necessary.

Reviewer 3: Especially if it is opt-in as it is in the suggested example it is easier to add it separately once the main formatting RFC has been accepted.

Reviewer 4: Sorry, I read your comment on that we should keep discussions in threads.
I thought I did the right thing, apparently I did not.

Reviewer 2: @SomeoneSerge Please always open a thread, so it can be resolved later! I'll inline [your comment](https://github.com/NixOS/rfcs/pull/166#issuecomment-1870697420) here:

> Hi, I just noticed that switching from `stdenv.mkDerivation {\n` to `stdenv.mkDerivation (finalAttrs: {\n` changes the indent of the body, generating a sizeable diff:
> 
> ```
> ❯ nixfmt << \EOF
> { stdenv }:
> 
> stdenv.mkDerivation (finalAttrs: {
>   dontUnpack = true;
>   dontBuild = true;
>   installPhase = "touch $out";
> })
> EOF
> { stdenv }:
> 
> stdenv.mkDerivation (
>   finalAttrs: {
>     dontUnpack = true;
>     dontBuild = true;
>     installPhase = "touch $out";
>   }
> )
> ```
> 
> Not sure if this has always been this way?
> 
> EDIT: This was probably the wrong place to report? Feel free to hide the message, and sorry about the ping

And @rhendric's [reply](https://github.com/NixOS/rfcs/pull/166#issuecomment-1870703763):

> > EDIT: This was probably the wrong place to report? Feel free to hide the message, and sorry about the ping
> 
> I hope not; I think this is good feedback for the spec. There doesn't seem to be a section governing multi-line contents of parentheses, and whether any elements can be included on the same line as the opening parenthesis.

Indeed such feedback is appreciated, just make sure that the formatting actually conforms to the specification (which I think is the case here).

Author: Continuing the discussion in https://github.com/NixOS/rfcs/pull/166#discussion_r1480357346

    if cond1 then
      foo
    else if cond2 then
      bar
    else
      baz
    ```

### Single-line common ancestor expression rule

For any two (sub-)expressions that are fully on a common single line, their smallest common ancestor expression must also be on the same line.

**Example**

```nix
# Bad, expressions cond and bar are fully on the same line,
# but their smallest common ancestor expression is the entire if-then-else, which spans multiple lines
if cond then foo
else bar

# Okay, cond, foo and bar have the if-then-else as a common ancestor expression,
# which is also fully on the same line
if cond then foo else bar

# Bad, due to function application precedence, the smallest common ancestor expression
# of foo and bar is `foo || bar baz`, which spans two lines
foo || bar
  baz
```

**Rationale**

This rule has turned out to be very practical at catching code that could be potentially hard to understand or edit.

### Line length

Lines have a soft length limit that doesn't count indentation.
There may also be a hard length limit that includes indentation.
String-like values such as strings, paths, comments, urls, etc. may go over the hard length limit.

> These limits should be configurable, and the soft length limit should default to 100 characters.

Reviewer 5: ```suggestion
These limits should be configurable, and the soft length limit should default to 80 characters.
```

I am curious to know how much this was already discussed. Finding the right compromise for line length is essential.
It ensures the readability of expressions (also on smaller screen sizes) but doesn't affect the filesize significantly. 
While it might be less controversial to choose a larger line length limit (although it is a soft one), 

**I propose that we choose a smaller limit (such as 80 characters) and follow other established conventions, such as those in rust, python, or javascript, who really dug into that problem already.**

If it is configurable anyway it should be more strict and people who don't like that can change it

Rustfmt states, like most other popular formatting tools.

> Lines should be at most 100 characters. **It's even better if you can keep things to 80.**

Also prettier states

> For readability we recommend against using more than 80 characters:

- nix only uses two '  ' spaces for indentation, which is relatively low. 
  > Why should we have a higher line limit if indentation takes less space?
  > Keeping lines short and concise is a good thing as other tools have already been discovered.
  
  
- [rust](https://rustc-dev-guide.rust-lang.org/conventions.html#:~:text=Lines%20should%20be%20at%20most%20100%20characters.)
- [python](https://peps.python.org/pep-0008/#:~:text=The%20Python%20standard%20library%20is,docstrings%2Fcomments%20to%2072).
- [javascript](https://prettier.io/docs/en/options.html)

Author: You can have a look for yourself: https://github.com/piegamesde/nixpkgs/pull/3

Opinions on line lengths vary wildly, from people who want to go back to 80, to people who say the code looks best at ~130. But so far I've seen mostly positive feedback on the change from 80 to 100.

Also note that Nix ist not Rust or Python or any other language. Due to its nature, code often times contains long strings. We need to account for that in our decisions.

Reviewer 6: I think a good rule of thumb would be half the horizontal width of a reasonably large monospaced font size (14-15) on a 16:9 monitor.

Reviewer 2: @Atemu Interesting heuristic, just counted for mine and got 98 characters for my screen and a comfortable font size! Though if you add vim sidebars and file trees and co. that shrinks..

Reviewer 2: Here's the diff on Nixpkgs when changing the default line limit from 100 to 80: https://github.com/piegamesde/nixpkgs/commit/e345d638168c066fae7e5bcfee414eae919712b1

Reviewer 7: > Here's the diff on Nixpkgs when changing the default line limit from 100 to 80: [piegamesde/nixpkgs@e345d63](https://github.com/piegamesde/nixpkgs/commit/e345d638168c066fae7e5bcfee414eae919712b1)

On my phone the shorter line length improves readability because fewer lines are wrapped.

Reviewer 5: > Here's the diff on Nixpkgs when changing the default line limit from 100 to 80: https://github.com/piegamesde/nixpkgs/commit/e345d638168c066fae7e5bcfee414eae919712b1

I definitely prefer the 80s limit. And would recommend it to the rfc decision. I know nix is not any other language but smaller limits exist for many reasons. Having large limits just shows that someone cannot format his code properly. It makes code less readable if you squish more expressions and statements into a single line.
- Having exceptions though for long paths or single line strings is fine i think. 

Controversial topic

Reviewer 2: Note that this is only the _default_ line width, it's specified to be configurable.

Personally I don't have a strong opinion on this, 80 sounds fine too.

Reviewer 8: I'm opposed to an 80 character limit, but in favor of an 80 character recommendation with a 100 character default limit.

However far beyond that, @hsjobeki I am strongly opposed to insulting people's skill on a code style generalization, and I also would appreciate it if you didn't hinder inclusivity with male-default language. Thank you.

Author: Please remember that this is about the format to be used by Nixpkgs. That's the value which matters to me, regardless of whether it ends up matching the nixfmt CLI default or not.

Also please remember that this is about an *automatic formatter*, the goal is precisely that nobody should have to manually format any code anymore.

Reviewer 9: I believe that a lot of newlines distract from the content of the code. Often a statement is one unit of logic and splitting it up makes it harder to follow what's really relevant.

Having fewer newlines packs the screen more densely with code so less scrolling is needed to get an overview and orientation.

80 character rules are very old from times with lower resolutions. Most people don't code on their phones.

I have also no very strong opinion on this but these are the arguments why I reallylike the change and advocate for keeping it.

Author: > - Having exceptions though for long paths or single line strings is fine i think.
> Controversial topic

[nope](https://github.com/piegamesde/nixpkgs/pull/4/commits/d6b557762ee83fb9cb4a56bf8f0fd3d01c791d2f)

----

Btw I'm currently experimenting with some more rules which allow to bypass the line length limit in some circumstances: https://github.com/piegamesde/nixpkgs/pull/4/files/eb3daa3..b247f67

There are some false positives, but overall I like the direction where this is going. (This will be discussed in the next meeting.)

Reviewer 2: Discussed in the meeting today, we agree that the default of a 100-char soft limit is pretty good. We came to that conclusion already before when we went through various limits. But in any case, it's explicitly stated to be configurable, so if people have different preferences, they can adjust it as necessary. We think this doesn't need to be discussed further.

Reviewer 10: Personally, I care about a hard limit much more than a soft limit. That way I can predictably avoid wrapping.

Author: @ursi You cannot avoid wrapping either way, because a lot of lines simply are longer than the line limit with no possibility of breaking them up.

- Compacting multiline arguments like this:
  ```nix
  #4b
  function arg1 {
    # stuff
  } arg3

  #5b
  function {
    # …
  } {
    # …
  }
  ```
  - This violates the guideline of the indentation representing the expression structure, and thus reduces readability.
  - This does not work well with line length limits on short arguments like in example #3.

### Function declaration

**Description**

- The body of the function is not indented relative to its arguments.
- Multiple ("simple") identifier arguments are written onto the same line if possible.
- Attribute set arguments always start on a new line; they are not mixed with identifier arguments.
  - If they have few attributes, the argument may be written on a single line, otherwise the expanded form is used.
- Attribute set arguments have their attributes on a new line each with indentation, followed by a trailing comma.

**Examples**

```nix
#1
name: value: name ++ value

#2
name: value:
{
  "${name}-foo" = value;
}

#3
{ pkgs }: pkgs.hello

#4
args@{
  some,
  argument,
  default ? value,
  ...

> }:

Reviewer 11: _I know this is a holy war, but my eyes and brain refuse to read such arguments without leading commas... And for a reason (described below)._

__Leading aligned commas serve additional information for quicker recognition by a human.__
You see a geometric structure quicker, where comma position is fixed at the "x" axis. This gives quicker information about whether each line specifies an input attribute and whether its syntax is correct.

```nix
{ some
, argument
, some-other
, default ? value
, ...
}@args:
```

With leading commas the algorithm to parse a single line of an attrset-argument syntax with eyes is as follows:
1. locate a vertical line of strokes (`,` `{` `}`) on the left
2. isolate a continuous set of characters apart from whitespace
3. parse an attribute name

![nix-rfc0166](https://github.com/NixOS/rfcs/assets/464913/dc8e1702-91b7-4e1f-b497-f61d27f4e2b8) ![nix-rfc0166-myscr_2023-11-17T03:51:48+04:00](https://github.com/NixOS/rfcs/assets/464913/888df73e-e408-40e3-af2e-e340e3a81eab)

Human brains are fast at tracing lines and highlighting words on a solid background. Just a single operation is needed to validate most of the attrset-argument syntax.

---

```nix
args@{
  some,
  argument,
  some-other,
  default ? value,
  ...,
}:
```

With trailing commas the algorithm to parse a single line of an attrset-argument syntax with eyes is as follows:
1. go to a line above
2. traverse it to the end
3. check whether the last character is either:
    - `{`
    - `,`
4. go back to a line below
5. traverse the current line to the end
6. check whether the last character is `,`
7. cut a continuous set of characters, which:
    - begins after ` `
    - ends before `,` (or also ` ` if trailing commas are not forced)
8. parse an attribute name

This not only applies to parsing, but navigating to the end of each line to write a single character if it's not in place.

Now lets see how people with poor vision can misinterpret the trailing comma pattern in Nix language:
![image](https://github.com/NixOS/rfcs/assets/464913/9a4ad4f9-c0f9-497b-8e72-8a991b4aedc9)
Is it a list like this?
```nix
[
  arg-checksum
  argument-cpufreq
  program
  chromium
]
```

But with leading commas you would never confuse it with a list:
![image](https://github.com/NixOS/rfcs/assets/464913/3b703044-1ed4-4de4-864a-a5354a8801a9)

---

Some benefits of leading commas:
1. No need to traverse each line to the end.
3. No need to check for scattered unaligned commas.
4. Simpler rule to discern attribute name (simply surrounded by whitespace).
5. Better distinction of attrset-arguments with other elements of the language (e.g. never confusing attrset-arguments with attrset definition, which uses visually similar `;` character).
6. Not breaking nixpkgs code base and originality, where almost each invocation of something like `nix edit nixpkgs#hello` shows the well-distinguished geometric pattern of package dependencies.

Yes, there are valid points for leading commas, mainly being diff-friendly. But everything cannot be ideal. Sacrificing so much of code readability for a single line diff benefit does not sound optimal, IMHO. The mentioned problems of editors, doing wrong folding of comments with leading commas, is a tooling problem and not a language formatting.
If the first line diff is considered so important, maybe add support for leading commas (similarly to trailing commas) like this:
```nix
{
, some
, argument
, some-other
, default ? value
, ...
}@args:
```

Author: My personal requirements for being okay with leading commas are:

1. Propose a good format that works with per argument comments, multiline comments, multiline arguments. Simply liking leading comments in the trivial cases is not sufficient.
2. Have support for leading comma on the first argument in the language

Another argument, albeit a more idealistic one, is that a format should be independent of the indentation size: leading commas would look weird for any indentation other than two space

Reviewer 5: If we take a look around,
barely any other language uses leading commas. Typing leading commas is thus unintuitive for most people coming from other languages. 

Haskellers have developed the idiom of leading comma.

Author: To add to that, all languages I know of in which leading comma is used, either don't or didn't support trailing commas on the last item. Leading commas are always a workaround as the less-bad option when there are no trailing commas.

Reviewer 2: We looked at this in today's shepherd meeting. We are convinced that the trailing commas are overall better, which also seems to align with most people's reactions in this thread, so we intend to keep the RFC as is.

We did however expand the argumentation in the RFC to also include the good new arguments given in this thread: https://github.com/nix-rfc-101/rfcs/commit/4f3b3ce5d704755d6527a9eaae9d01b9b430104d

Reviewer 11: ### trying to structure comments in this thread

So, the comments above with criticism of the current leading style formatting can be structured into these theses:

1. leading commas style doesn't have a complete proposal with enough cases documented
2. leading commas are _unintuitive_ or unusual comparing to other languages
3. leading commas are a historical workaround _("less-bad option when there are no trailing commas")_
4. trailing commas seem to _"align with most people's reactions in this thread"_

I think thesis [1] is objective and valid. And a complete proposal has to be defined. I try to address some points below. Yes, it's not always trivial, with compromises in some cases. But there is no ideal solution with trailing commas either.

As for thesis [2] and [3] it's hard to argue, because being quite subjective.
Except the comparison with other languages, where we can admit statistical difference, but not the outcome it brings. Is being similar always a good point or being different always a bad point?

As for the thesis [4] the situation is not as clear as stated (@infinisil):
![image](https://github.com/NixOS/rfcs/assets/464913/3ee3cec5-36ae-4191-aa24-fe22edc40afa)
Even here we have 50/50%. Many contributors do not participate in reading RFC 0166, do not vote. And I think such a drastic change in nixpkgs formatting may come as a shock to most.

---

## a few more arguments in favor of leading commas

I am very sorry for not providing exact subjective statements this time below, but, IMHO, these are worth your attention too :pray:.

### unusual solutions to unusual challenges

> barely any other language uses leading commas

I think our answer to this:
_barely any other language solves problems which Nix is designed to solve_.

I believe it's **not a coincidence** to have decisions which look weird compared to other tools.

Other conventional languages are not used to **de-sugar input attribute sets** (say structures/records) to this extent as Nixpkgs does in practice. Leading commas is one of the most used syntax patterns, since we deal with **lots of input entities** in a single function most of the time. I cannot believe we can trade the readability of the cornerstone input Attribute Set language construct for one line diff, unadapted editors, etc.

Nix ecosystem is on the rise. Lets stabilize things which are not blockers and concentrate on those which are (e.g. flakes). Leading commas did not stop us from creating the biggest software repository on the Earth.

### Nixpkgs historical case

Forcing Nixpkgs to get rid of leading commas is the biggest break to the part of its established syntax style. I would say it's the Nixpkgs mark, which you see in each package definition. Out of all syntax patterns in the RFC the leading commas pattern is the most used and the most stable one in Nixpkgs (and not only there). The debate is kinda similar to the established `nix shell` naming, which has polished through years (losing which would mean a huge loss for our community (https://github.com/NixOS/nix/issues/4715#issuecomment-1002161350)).

### first argument diff in nixpkgs

Usually, package is dependent on stdenv or lib, so these are placed first.
If we take this as a rule, most first argument diffs will be neutralized. Surely, this is a compromise as it covers top-level package functions mostly. But almost any part of the decision to take on formatting will be a compromise regardless of the comma position.

### tooling

Still, implementing a language support for leading comma on the first argument is most likely an overkill.
**Tooling is the key** in my opinion. Lets add support for established format to editors, rather than trying to adapt code of one of the largest and most active repositories in the world to them at this enormous cost.

### let's briefly go through the points of criticism of leading commas style in RFC

* _"Moving items around with this style may require editing lines."_
    True with the first item. Replacing letter `{` with `,` and vice versa. 
    An editor plugin could automate this if it's a problem. Such diffs will be rare in nixpkgs.
* _"Inconsistent with the [expression expansion guidelines](https://github.com/NixOS/rfcs/blob/463a858e57e4d68f4bfb1d38f20d8ac852ac4574/rfcs/0166-nix-formatting.md#expansion-of-expressions), which disallows forms like { some; some should start on a new line instead."_
    As I see, the core idea of the guidelines regarding "_each item must be on its own line_" is the ease of items begin search and vertical space consistency. Leading commas are consistent in having special character (`{` or `,`) in the beginning of each line.
    The current wording, however, does not strictly prevent other symbols to appear on the same line. Except the counter-example in the subsection of the rule: _"e.g. [ firstElement in a multi line list is not allowed"_.
    So, just a slightly different wording for this rule will be:
    * each line must contain at most one item
    * all lines must begin vertically aligned with the same type of character (e.g. letter, special)
    * _(additionally)_ all items must begin vertically aligned between each other
* _"This does not work well with leading @ bindings."_
    Use trailing @ bindings. What are the downsides?
* _"It's unclear whether comments belong to the next or the previous argument."_
    Rule: comment (if any) must start before argument's leading comma.
* _"The leading comma style was a lesser-evil workaround for the lack of trailing commas in the Nix language..."_
    In my opinion, this point is only a historical one, without any influence on our current decision process. Not all available features in the language need to be used.

### testbed

> For any proposal, please use https://github.com/NixOS/nixpkgs/blob/a1ed34a3161eb27eab6c743a54ce0dbbe6db0b21/pkgs/stdenv/generic/make-derivation.nix#L105-L203 as a benchmark. It covers almost all relevant corner cases

Nice one! It is quite a good example of how leading commas help to orient among many attribute set arguments. Personally, I like most of it and just need to formalize and standardize the decisions made there to address:
> format that works with per argument comments, multiline comments, multiline arguments

@piegamesde, btw, is there already a sample of this benchmark, formatted in the proposed trailing commas style?

---

_I hope I did not offend anybody :pray:. I highly appreciate the work of all the contributors and I'm very grateful to Nix community in general for doing such a great piece of tech and giving me an opportunity to have a Nix job._

Author: To me, your argument in favor of leading commas boils down to "we can do this". Yes, absolutely. All of these issues are solvable and might have a decent solution. We are making the rules right now, we absolutely can change these rules if we can think of better ones.

But if you take a step back, and ignore all historic path dependencies and any personal "I just prefer it like this" preferences, and have a look at both options and neutrally compare them—do you still think that having leading commas in function declarations is preferable over trailing commas? At least to me, the choice is pretty easy: One of them comes with lots of special cases and caveats and potential inconsistencies and special requirements for tooling support, the other one is relatively straightforward.

-------

You can have a look at the current Nixmft output here: https://github.com/piegamesde/nixpkgs/blob/nixfmt-2/pkgs/stdenv/generic/make-derivation.nix#L76-L182. Keep in mind that the formatter is constrained by the input in a way a human adapting the code would not be, and also that any vertical alignment got removed in the process, which is a different topic. After some minor manual cleanup, it could look like this (note that I chose to reintroduce the vertical alignment, to make it closer to the original, so that we may better ignore that aspect in the comparison):

```nix
  mkDerivationSimple =
    overrideAttrs:

    # `mkDerivation` wraps the builtin `derivation` function to
    # produce derivations that use this stdenv and its shell.
    #
    # See also:
    #
    # * https://nixos.org/nixpkgs/manual/#sec-using-stdenv
    #   Details on how to use this mkDerivation function
    #
    # * https://nixos.org/manual/nix/stable/expressions/derivations.html#derivations
    #   Explanation about derivations in general
    {
      # These types of dependencies are all exhaustively documented in
      # the "Specifying Dependencies" section of the "Standard
      # Environment" chapter of the Nixpkgs manual.

      # TODO(@Ericson2314): Stop using legacy dep attribute names

      #                           host offset -> target offset
      depsBuildBuild ? [ ], #              -1 -> -1
      depsBuildBuildPropagated ? [ ], #    -1 -> -1
      nativeBuildInputs ? [ ], #           -1 ->  0  N.B. Legacy name
      propagatedNativeBuildInputs ? [ ], # -1 ->  0  N.B. Legacy name
      depsBuildTarget ? [ ], #             -1 ->  1
      depsBuildTargetPropagated ? [ ], #   -1 ->  1

      depsHostHost ? [ ], #                 0 ->  0
      depsHostHostPropagated ? [ ], #       0 ->  0
      buildInputs ? [ ], #                  0 ->  1  N.B. Legacy name
      propagatedBuildInputs ? [ ], #        0 ->  1  N.B. Legacy name

      depsTargetTarget ? [ ], #             1 ->  1
      depsTargetTargetPropagated ? [ ], #   1 ->  1

      checkInputs ? [ ],
      installCheckInputs ? [ ],
      nativeCheckInputs ? [ ],
      nativeInstallCheckInputs ? [ ],

      # Configure Phase
      configureFlags ? [ ],
      cmakeFlags ? [ ],
      mesonFlags ? [ ],
      # Target is not included by default because most programs don't care.
      # Including it then would cause needless mass rebuilds.
      #
      # TODO(@Ericson2314): Make [ "build" "host" ] always the default / resolve #87909
      configurePlatforms ?
        lib.optionals (stdenv.hostPlatform != stdenv.buildPlatform || config.configurePlatformsByDefault)
          [
            "build"
            "host"
          ],

      # TODO(@Ericson2314): Make unconditional / resolve #33599
      # Check phase
      doCheck ? config.doCheckByDefault or false,

      # TODO(@Ericson2314): Make unconditional / resolve #33599
      # InstallCheck phase
      doInstallCheck ? config.doCheckByDefault or false,

      # TODO(@Ericson2314): Make always true and remove / resolve #178468
      strictDeps ?
        if config.strictDepsByDefault then true else stdenv.hostPlatform != stdenv.buildPlatform,

      enableParallelBuilding ? config.enableParallelBuildingByDefault,

      meta ? { },
      passthru ? { },
      # position used in error messages and for meta.position
      pos ? (
        if attrs.meta.description or null != null then
          builtins.unsafeGetAttrPos "description" attrs.meta
        else if attrs.version or null != null then
          builtins.unsafeGetAttrPos "version" attrs
        else
          builtins.unsafeGetAttrPos "name" attrs
      ),
      separateDebugInfo ? false,
      outputs ? [ "out" ],
      __darwinAllowLocalNetworking ? false,
      __impureHostDeps ? [ ],
      __propagatedImpureHostDeps ? [ ],
      sandboxProfile ? "",
      propagatedSandboxProfile ? "",

      hardeningEnable ? [ ],
      hardeningDisable ? [ ],

      patches ? [ ],

      __contentAddressed ?
        (!attrs ? outputHash) # Fixed-output drvs can't be content addressed too
        && config.contentAddressedByDefault,

      # Experimental.  For simple packages mostly just works,
      # but for anything complex, be prepared to debug if enabling.
      __structuredAttrs ? config.structuredAttrsByDefault or false,

      env ? { },

      ...
    }@attrs:
```

Reviewer 4: why do this
```nix
{ foo
, bar
, foobar
}:
```
but not this?
```nix
{ foo = "foo"
; bar = "bar"
; foobar = "foobar"
;}
```

Author: @quantenzitrone indeed, both syntax constructs are related. What you point out is precisely the converse of our current rules: *Because* leading `;` look so bad, we also forbid leading `,` for the sake of consistency.

By the way, I am not opposed to leading item separators instead of trailing ones. However, my requirement for this would be to allow a leading `,` on the very first item. (Then, this style would be equivalent to simply having a bullet point list except that it starts with `,` instead of `-`)

Author: > - on the contrary, with leading commas the solution is simple:
>   - align `,` on the left
>   - align `#` on the right

@AleXoundOS I'm not sure what exactly you mean by this, could you please expand and maybe also provide some examples?

Reviewer 11: @piegamesde, I mean this:

### trailing commas unaligned

```
  nativeBuildInputs ? [ ], #           -1 ->  0  N.B. Legacy name
  propagatedNativeBuildInputs ? [ ], # -1 ->  0  N.B. Legacy name
  depsBuildTarget ? [ ], #             -1 ->  1
  depsBuildTargetPropagated ? [ ], #   -1 ->  1
```

### trailing commas with `#` aligned

```
  nativeBuildInputs,           # -1 ->  0  N.B. Legacy name
  propagatedNativeBuildInputs, # -1 ->  0  N.B. Legacy name
  depsBuildTarget,             # -1 ->  1
  depsBuildTargetPropagated,   # -1 ->  1
```

### leading commas with both `,` and `#` aligned

```
, nativeBuildInputs           # -1 ->  0  N.B. Legacy name
, propagatedNativeBuildInputs # -1 ->  0  N.B. Legacy name
, depsBuildTarget             # -1 ->  1
, depsBuildTargetPropagated   # -1 ->  1
```

### trailing commas with both `,` and `#` aligned (weird, IMHO)

```
  nativeBuildInputs          , # -1 ->  0  N.B. Legacy name
  propagatedNativeBuildInputs, # -1 ->  0  N.B. Legacy name
  depsBuildTarget            , # -1 ->  1
  depsBuildTargetPropagated  , # -1 ->  1
```

Reviewer 4: 2, 3 and 4 would not work with the vertical alignment nonpreservation rule

https://github.com/nix-rfc-101/rfcs/blob/463a858e57e4d68f4bfb1d38f20d8ac852ac4574/rfcs/0166-nix-formatting.md?plain=1#L437

It may be good to specifically point out that vertical alignment of comments is not preserved as well.

Either there or in the section about comments: https://github.com/nix-rfc-101/rfcs/blob/463a858e57e4d68f4bfb1d38f20d8ac852ac4574/rfcs/0166-nix-formatting.md?plain=1#L590C1-L607

Reviewer 2: We've discussed this again in the shepherd team but still stand by our decision to use trailing commas. Thanks for your input regardless though!

Reviewer 11: @infinisil, thank you for paying attention to this! Let's see how it goes. But, even considering the current spontaneous votes status ![image](https://github.com/NixOS/rfcs/assets/464913/9e13a1cd-75c0-4929-a626-698f740e679e) (under the first message of this thread), 54% does not seem to be enough for a radical change to happen and community debate seems to continue as change begins.

Author: This is not how our decision making process works.

Author: *TL;DR: Everything I've already said before and also written into the RFC, but in more detail and with many more words.*

## Leading commas

Okay, so let's design a format for function arguments and leading commas and try to make it as good as possible. I'll cover all relevant design questions. (More may come up during an actual implementation attempt, but this should cover most of the ground.)

### 1. Basics

Let's start simple: Just three arguments, all single-line. There are already two possibilities to choose from.

```nix
# 1
{
  arg1
, arg2 ? null
, arg3
}:

# 2
{ arg1
, arg2 ? null
, arg3
}:

[]
```

You may say "obviously we want the second one", but this one contradicts our current expansion rules:

> - If expanded into multiple lines, each item must be on its own line.
>   - This also applies to the first item, so e.g. `[ firstElement` in a multi line list is not allowed.

I agree though that the second style is preferable, so now another choice has to be made: Adapt the other rules derived from or accept the inconsistency.

```nix
[ first
  second
  third
]

{ one = "1";
  two = "2";
  tree = "3;
}
```

Taking that style further quickly leads to chaos and further inconsistencies:

```nix
rec {
  one = "1";
  two = "2";
  three = "3";
}

function call [
  first
  second
  third
]

function call # long line
  [ first
    second
    third
  ]
```

This may even have ripple effects onto other syntax constructs like parentheses. So it looks like making only one exception to the expression expansion rule is the less bad option here.

### 2. Multiline arguments

Things are already getting tricky.

```nix
# 1
{ one ? [
    1
    1
  ]
, two ?
    if "two" == 2 then 2 else "two"
, three ? function [
    1
    2
  ]
}:

# 2
{ one ? [
  1
  1
]
, two ?
    if "two" == 2 then 2 else "two"
, three = function [
  1
  2
]
}:

# 3
{ one ?
  [
    1
    1
  ]
, two ?
  if "two" == 2 then 2 else "two"
, three ?
  function [
    1
    2
  ]
}:

# 4
{ one
  ? [
    1
    1
  ]
, two
  ? if "two" == 2 then 2 else "two"
, three
  ? function [
    1
    2
  ]
}:

[]
```

There are many other variations to handle the indentation here, but it quickly becomes clear that the base indent must be at +2 of the start of the function argument name, or –equivalently– at +4 of the start of its line. This is in violation of our "double indentation" rule, which states that from one line to the next one may only indent upt to a single level. Currently, that rule is the only "hard" rule in the RFC style, with absolutely zero exceptions.

The only one which avoids this to be somewhat taken seriously is option \#4, but it is as desperate an attempt as it looks. It has several downsides, from diff noise when switching between multi-line and single-line defaults to weird interactions with function calls (not shown in this example).

So I'll accept defeat and make one exception to the sacred double indentation rule, and proceed with style \#1.

### 3. `@`-pattern

In Nix, you can have the `@`-pattern either before or after the argument destructuring. Both are equivalent.

```nix
# 1
args@{
  one
, two
, three
}:

# 2
{ one
, two
, three
}@args:

# 3
{ one
, two
, three
}
@args:

# 4
{ one
, two
, three
}
@ args:

# 5
{ one
, two
, three
} @ args:

[]
```

Currently in Nix leading `@` is the more prevalent convention (estimated through some quick greps through `nixpkgs`), but clearly it doesn't work here: the first argument without a leading comma just looks too inconsistent.

So the only sensible option is to normalize them to the trailing version during formatting. (Note that this is not as trivial as it may look like at a first glance, since any token may have comments associated with it which need proper handling as well. But it should be doable.)

There are lots of way about how to handle the whitespace for trailing `@`-patterns, but on a cursory glance this looks like a mostly subjective decision so I'll arbitrarily pick style \#4.

### 4. Comments

Arguments may have comments associated with them. These may come in various forms, `#`, `/*`, `/**`, single-line or spanning multiple lines. Trailing comments on a line are trivial and don't need to be discussed.

```nix
# 1
{ /* comment */
  one
, /*
    multiline
    comment
  */
  two
, # comment
  three
, # multiline
  # comment
  four
}:

# 2
/* comment */
{ one
/*
  multiline
  comment
*/
, two
/* comment */
, three
# multiline
# comment
, four
}:

[]
```

Style \#1 has the disadvantage that the comma gets increasingly distanced from the argument it is associated with (comments may be long). It also loses the clear vertical line to the left which is one of the big advantages of a leading comma style.

Style \#2 looks more promising though. When using `#` comments it even has a clear vertical line for visual reference. It may be tempting to thus decide to normalize all `/*` comments to `#` for that purpose. But this may be a controversial change on its own, but more importantly this cannot be done for `/**` doc comments according to RFC 145. The RFC does not specify any `#`-starting doc comments, so this idea is out. Additionally, style \#2 has the issue that the first comment collides with a potential comment for the entire function. This is not compatible with RFC 145 either.

Therefore, using style \#1 for comments is the only viable option here.

## Trailing commas

Okay and now let's do the same thing again, but with trailing commas.

### 1. Basics

Same decision as last time:

```nix
# 1
{
  arg1,
  arg2 ? null,
  arg3,
}:

# 2
{ arg1,
  arg2 ? null,
  arg3,
}:

[]
```

This time it is easy though, we have no problem in following the expression expansion rule and thus pick style \#1.

We also have the decision of whether or not we want to put a comma after the last argument. But not having a trailing comma on the last argument would be even worse than the leading comma style, remember that's how we got into this situation in the first place! (The leading comma style was a lesser-evil workaround back in the days where the language didn't allow a trailing comma on the last item.)

### 2. Multiline arguments

There is little to debate here. The syntax looks so similar to bindings in attrsets, that we can simply copy the rules over 1:1 (replacing `=` with `?` and `;` with `,`). Yay for consistency!

```nix
# 1
{
  one ? [
    1
    1
  ],
  two ?
    if "two" == 2 then 2 else "two"
  three ? function [
    1
    2
  ],
}:

[]
```

### 3. `@`-pattern

Our first non-trivial decision with trailing commas.

```nix
# 1
args@{
  one,
  two,
  three,
}:

{
  one,
  two,
  three,
}@args:

# 2
args @ {
  one,
  two,
  three,
}:

{
  one,
  two,
  three,
} @ args:

# 3
args@
{
  one,
  two,
  three,
}:

{
  one,
  two,
  three,
}
@args:

# 4
args @
{
  one,
  two,
  three,
}:

{
  one,
  two,
  three,
}
@ args:

[]
```

Leading `@`-binding looks better than the trailing one, but it is more common in Nixpkgs anyways. We could try to normalize to always use leading `@`-bindings, but if possible this should be avoided so let's discard that option.

For consistency, the leading and trailing version should look the same. There are many possibilities on how to place the whitespace here. I'll optimize for the more common leading `@` style and pick option \#1 or \#2.

### 4. Comments

Nothing really to discuss here either. Again, we'll just copy the style over 1:1 from bindings.

```nix
# 1
{
  /* comment */
  one,
  /*
    multiline
    comment
  */
  two,
  # comment
  three,
  # multiline
  # comment
  four,
}:

[]
```

## Conclusion

Yes, we can somehow get leading commas to work. But at every corner we are faced with tough decisions between sub-optimal options, having to make exceptions to rules and allowing inconsistencies. The only argument in favor of leading commas would be the widespread usage in current Nixpkgs. However, the RFC states on this topic:

> When deciding between two *equally good* options, currently prevalent formatting style in Nixpkgs should be followed. The emphasis here is on "equally good". We should not fear making radical changes to the current style if there are sufficient arguments in favor of it.

As we've just seen, both options are not "equally good". To me, trailing commas are good but leading commas are *okay at best*.

And that's the reason why the trailing commas style was chosen in this RFC.

{
  # body
}

#5
{ pkgs }:
name: value: 
{
  # body
}
```

**Rationale and alternatives**

- Have leading commas for parameters in attribute set arguments, like currently done in Nixpkgs
  ```nix
  #6
  { some
  , arg
  }:
  #7
  args@{
    some
  , argument
  # Single line comment
  , commentedArgument
  , # Comment on the value
    # multiline comment
    default ? value
  , ...
  }:
  # …
  ```

>   - This leads to problems with the first argument, as leading commas are not allowed. `{ some` is discouraged by the the style guidelines; `some` should start on a new line instead. Also, this does not work well with `@` bindings.

Reviewer 12: ```suggestion
  - This leads to problems with the first argument, as leading commas are not allowed. `{ some` is discouraged by the style guidelines; `some` should start on a new line instead. Also, this does not work well with `@` bindings.
```

- If the operand can also fit on the same line as the operator, it's put there
- Otherwise, the operand usually starts indented on a new line, with special handling for parenthesis, brackets, braces, function applications

Operator chains may be compacted as long as all lines between the first and last one are indented.

**Examples**

```nix
# These chained associative operators have increasing precedence, so they're _not_ treated the same
foo
-> # <- The operator starts on a new line, but right operand is all of the below lines, they don't fit here, so indent
  bar
  ||
    baz
    && qux # <- The operand fits on this line

# These chained associative operators have decreasing precedence, so they're treated the same
foo
&& bar # <- All of these operands are just identifiers, they fit on the same line
|| baz # <- We shouldn't indent these lines, because it misleads into thinking that || binds stronger than &&
-> qux

[
  some
  flags
]
++ ( # <- The operator is on a new line, but parenthesis/brackets/braces can start on the same line
  foo
)
++ optionals condition [ # <- As are multiline functions applications
  more
  items
]
++

>   runCommand name # <- But only functions whose last argument can start on the same line

Reviewer 12: Whose last argument is what?

Reviewer 2: Should be

> But only functions whose last argument starts on the same line

Reviewer 12: ```suggestion
  runCommand name # <- But only for functions whose last argument starts on the same line
```

    ''
      echo hi
    ''
    test

# We can use a more compact form as long as all inbetween lines are indented
bar // {
  x = 10;
  y = 20;
} // baz

# Bad, we can't use the more compact form because an intermediate line is not indented.
{
  x = 10;
  y = 20;
} // foo // {
  z = 30;
  w = 40;
}

> ```

Reviewer 12: That's not clear to me intuitively, and only makes sense if you follow the "intermediate lines must be indented" rule literally. A "good" alternative would help here. Is that okay, since sequenced operators start on the same level?

```nix
{
  x = 10;
  y = 20;
} // foo
// {
  z = 30;
  w = 40;
}
```

It doesn't feel right.

Reviewer 2: Ah we just pushed something earlier that made this not quite work out. Notably this compact form should be specific to binders. I undid it again and added another example in https://github.com/NixOS/rfcs/pull/166/commits/940312fbdf41a89e4fdd3756f3724bc9cabd1d6f.

We think that making operators be on the same indentation level would be most consistent, but the compact form is there to make cases like these still work:

```nix
{
  foo = foo // {
    # Only 1 extra indentation level here
    bar = 10;
  };
}
```

The intermediate indentation requirement is there so it's easier to visually see where the binder starts and ends, you can scan a column to find binders that way.

if
  matches != null
  && builtins.length matches != 0
then
  { inherit path matches; }
else if path == /. then
  null
else
  go (dirOf path);
```

**Rationale and alternatives**

- Attribute sets and lists could start on the same line as the if keywords, saving an indentation level on their body:
  ```nix  
  #1a
  if builtins.length matches != 0 then {
    inherit path matches;
  } else if path == /. then [
    1
    2
  ] else
    go (dirOf path);
  ```
  - This results in inconsistent vertical start of the keywords, making the structure harder to follow
- Have the `then` on the start of the next line, directly followed by the if body:
  ```nix
  #1b
  if builtins.length matches != 0
    then { inherit path matches; }
  else if path == /.
    then [
      1
      2
    ]
  else go (dirOf path);

  #1c
  if builtins.length matches != 0
  then { inherit path matches; }
  else if path == /.
  then [
    1
    2
  ]
  else go (dirOf path);
  ```

### with, assert

**Description**

- The body after the statement starts on a new line, without indentation.
- There may be exceptions for common idioms, in which the body already starts on the same line.

**Examples**

```nix
with pkgs;

> assert foo == bar; {

Reviewer 12: This sample is in violation of the firm rule, and if it's intended it doesn't say why it's a "maybe" exception.
```suggestion
assert foo == bar;
{
```

One exception could be

```nix
with pkgs; [
  foo
  bar
]
```

Reviewer 2: > This sample is in violation of the firm rule

The word "firm" doesn't appear in the RFC, which rule do you mean?

Reviewer 12: Ah, I meant the corresponding rule was stated firmly, no maybe's.

  meta.maintainers =
    with lib.maintainers; [
      some
      people
    ];
}
```

### let

**Description**

- Let bindings are *always* multiline.
- The "let" part is indented one level, but not the "in" part.
- Each item in the "let" part is indented and starts on its own line.
  For more details, see the [binders section](#binders).

> - The "in" part starts on a new line.

Reviewer 12: Is there a reason we can't smooth that down for expressions that fit in a line, so we can have something like

```nix
let
  # lots of stuff
in result
```

Reviewer 2: The motivation behind always having a newline is for easier editing and better diffs. E.g. say you have

```nix
let
  foo = 10;
in foo
```

And you want to insert a function argument (or another `let in` or an `assert`/`with`), you'd have to go inbetween `in` and `foo` to do that, and the diff would be

```diff
 let
   foo = 10;
-in foo
+in
+bar:
+foo
```

In the current proposal it would be just

```diff
 let
   foo = 10;
 in
+bar:
 foo
```

However, we need to consider trade-offs, I have the feeling that it might be worthwhile to allow no newline after `in`.

Reviewer 12: Minimal diffs are generally good, but as you say one has to consider it in context and weigh diff readability against code readability.

Author: The reason why I'd be against the compact form here is that this kind of looks weird in combination with most statements but simple identifiers (`in if`, `in foo: bar`. So for consistency's sake it's easier to always have a newline. But if you think that simple identifiers look too weird on their own line, we could special case just that

Reviewer 13: Really not a fan of not indenting the inner expression after the `in` but still putting it on a new line. To me this feels very wrong, like formatting the following in C:

```c
return
some_method_call(
  foo,
  bar
);
```

instead of

```c
return some_method_call(
  foo,
  bar
);
```

IMHO always attaching the `in` to the first line of the inner expression looks and reads better:

```nix
let
  x = 1;
in 2 * x

let
  x = 1;
in [
  x
  (x + 1)
  (x + 2)
]

let
  foo = bar;
in stdenv.mkDerivation {
  ...
}
```

And generally I think the same for attaching the `let` to the end of the last line of the outer expression for the same reason:

```nix
{
  stdenv,
}: let
  foo = ...;
in stdenv.mkDerivation {
  ...
}
```

Reviewer 12: From what I've seen in @infinisil's code contributions to Nixpkgs, the motivation behind the unindented newline pattern could be to encourage using `let` blocks to declare the provenance of symbols at the top of a file (the antithesis to `with pkgs; with lib;`). You'd usually leave an empty line between that and the business logic. Since it's a homogeneous rule you can repeat that pattern in nested scopes. I think it's quite nice.

Reviewer 2: In the meeting today we reached general agreement that it's okay to always have a newline after `in`

Reviewer 2: > Can we consider forcing a newline after `in`? The examples above read somewhat poorly to me. I think `in stdenv.mkDerivation {` reads best

I'm not sure what you're arguing for here, you can't both force a newline after `in` and then show an example where that's not the case :)

Reviewer 14: Right. What I meant to say is, every example that isn't `in stdenv.mkDerivation {` read badly to me. So I'm thinking if forcing a newline after `in` would be a net positive, considering that there's only one example of it being a little worse.

Author: You are mistaken about the proposed style here. In combination with other rules, you get the following instead:

```nix
  config =
    let
      foo = "bar";
    in
    mkIf (this.enable) {
      services.foo.bar = foo;
    };
```

(As you already said, that is a tradeoff between an additional indentation level for the content and keeping the top-level indent tidy)

Reviewer 2: We discussed this in the meeting again and added the mentioned alternatives to the RFC now: 6a6efcd2b5588ba662004b48314204628487f551

**Examples**

```nix
let
  foo = "bar";
in
if foo == "bar" then
  "hello"
else
  "world"
```

### Attribute sets and lists

**Description**

- Brackets and braces are generally written with a space on the inside, like `[ `, ` ]`, `{ ` and ` }`.

>   - Empty lists and attribute sets are written as`[ ]` and`{ }`, respectively.

Reviewer 12: ```suggestion
  - Empty lists and attribute sets are written as `[]` and`{}`, respectively.
```

This is a personal plea. Formatting rules should be practical even without tooling. On QWERTY without brace completion, and assuming regular typing ability, it's really annoying to type `Shift+[`, `Space`, `Shift+]` to express `{ }`. Without the space, you hold `Shift` and roll your fingers over `[` and `]`. Somewhat minor: If you have the formatter running automatically it will let the code jump needlessly on this trivial thing.

Arguably we should optimize for the still average case of QWERTY keyboards. Everyone else will know what they're doing anyway.

Reviewer 15: nit: Technically for the vanilla vim keybinds you can do `s`, `Enter`, `Tab`.
But I fully agree with you, and never leave a space in there as I don't really see the point.

Reviewer 12: > Technically for the vanilla vim keybinds you can do `s`, `Enter`, `Tab`.

TIL

Reviewer 2: This is something I really don't think matters a lot. We concluded that `{ }` is more consistent with `{ foo = 10; }`, so we chose that. I don't think anybody would be annoyed by this in practice.

Reviewer 6: > On QWERTY without brace completion, and assuming regular typing ability, it's really annoying to type `Shift+[`, `Space`, `Shift+]` to express `{ }`. Without the space, you hold `Shift` and roll your fingers over `[` and `]`.

Just hold shift the entire time. `Shift + Space` enters a space char. Only weird layouts like Neo differentiate here.

> Even more minor: When expanding a list or attrset with Vim-style controls, navigating to the space and hitting `r`, `Enter`, `i`, `Tab` is one interaction more than `w` to the first brace `a`,`Enter`,`Tab`.

In properly configured editors, you'd simply enter insert mode somewhere between the braces and press enter. It should automatically remove possible superfluous spaces at the end of the line and indent on enter.

This is an editor configuration problem, not a formatting one IMHO.

Reviewer 12: > In properly configured editors

If you're targeting software professionals, that's what you can assume. If you want to be inclusive to a wide range of people from different backgrounds (and we've seen some reports where the Nix language was the first encounter with programming), you have to take into account that it's not a given.

Again, not fighting here, just raising considerations I'd like people participating in this discussion to be aware of. RFC authors feel free to resolve close this thread at will.

Reviewer 16: While I don't find the ease of typing argument strong (though, it's likely that I'm biased as a "software professional"), I do think that requiring a space for empty attrsets and list looks strange, especially if you consider that probably most Nix users will have background in other programming languages. In my experience with various mainstream languages, formatters prefer `[]` and `{}` over `[ ]` and `{ }`.

I did some quick testing online and C# was the only exception:
* `dotnet format` (C#) - prefers `{ }` - [link](https://replit.com/@PetarKirov/csharp-formatting-empty-objects#main.cs)

other language formatters:

* Black (Python) - prefers `[]` and `{}` - [link](https://black.vercel.app/?version=stable&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4ACOAFJdAD2IimZxl1N_WlPS8NIZ-x4TNI2_ZJeJE5zxHZZWqBkdWPX0qG2H3hsDb2YmAGRmmWBxViohAyw5oXP2m5Fm6G0jZff5LMsRaXvuQ4O38I8uXQAAAAAzOT-uPfRThAABbo8BAAAA0fqherHEZ_sCAAAAAARZWg==)
* Prettier (JavaScript, TypeScript) - prefers `[]` and `{}` - [link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEBDATrzAXkwG1NyBdcgbgB0pIoNMIAjAKyM2HMwF8aIADQgIABxgBLdMlB5cEAO4AFPAjTIQ2ADaLsATw0jWubGADWcGAGVsAWzgAZSVDjIAZjrRxjpi1esxMxcAc2QYXABXHxA4O1Y4ABNEpMdsKBDI7BC4ADEIXDtsGCkMzWxImAhhEAALGDttAHVayXg0ILA4a3U2yQA3Nv1NMDQjEBdvXBhlUxCijy8Y9jQAD2tQ7TgARUiIeEXtbxEg3CnNVmwE7RqxXBcYJslEmFrkAA4ABhOFbybTMSaO5wKb9NwiACOe3gs3EGhQ2DQAFpXEkkjVcHAoZJMbNsgskJ4jjFvHZJOEoiTNnAAIIle6sSpwZRwXDOVyHY4gNDU3b7NyEpYiGBXJ4vN5IABMwtMkm0oQAwhA7ATYmgAKw1SLeAAqV3hRK5-WiAEkoClYNYwPcJDTzdYYPotpy4Hw+EA)
* Clang Format (C/C++) - prefers `{}` - tested with [this code](https://gist.github.com/PetarKirov/978c7d5ec2ca5338c762ec1cb4ec254d) at https://clang-format-configurator.site/
* `google-java-format` (Java) - prefers and `{}` - https://codebeautify.org/javaviewer/y23b1bfbb
* `dart format` (Dart) - prefers `[]` and `{}` - https://dartpad.dev/?id=50af8512ea53ee32ec05e995ebb58983
* `go fmt` (Go) - prefers `{}` - https://go.dev/play/p/9m0Zq_oz1ZO
* Ormolu (Haskell) - prefers `[]` - (tested [this code](https://gist.github.com/PetarKirov/924b09cb75eb2b87089512171501df54) on https://ormolu-live.tweag.io/)
* Rustfmt (Rust) - prefers `[]` and `{}` - [link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=dfbd5ebcc73c37c827cc271ebba03d49)
* Swift  - prefers `()`, `{}`, `[]`, and `[:]` - [link](https://swiftfiddle.com/cfsqjnuayzh55fxxi6ygx2hh7y)

Author: I personally don't have any overly strong opinions on this either way. However, for me consistency is important here. If we write empty lists without a space, then we should also write non-empty lists without: `[foo bar]`. Same for attribute sets. (This would be more consistent with function calls too btw.)

The space version was initially chosen because in the absence of strong arguments, current formatting conventions are followed.

Reviewer 2: In the meeting today, the shepherds agreed to consider the style without spaces, we'll post a Nixpkgs diff soon!

Reviewer 4: I'm also for no spaces, but probably only because I am used to it through alejandra.

Reviewer 6: I'm not a fan of no space. Lists in Nix use whitespace to separate items; either space chars or newlines. Whitespace therefore has great importance in these contexts.

```Nix
[a b c]
```

is inconsistent with

```Nix
[
  a
  b
  c
]
```

because there's suddenly whitespace between `[` and `a` where there was none before.

Additionally, I don't agree with the argument that other languages do no spaces as almost all "other languages" have "physical" separators between list items (usually `,`).

With `[a, b, c]`, it's a lot more bearable to have the `[` so close to the `a` than it is with `[a b c]` because "physical" characters in close proximity to the items is the norm.

***

As for `[]` vs `[ ]`, I feel like the whitespace further solidifies the emptiness of the list. `[]` reminds me too much of C arrays where the length is anything but clear.

Empty sets should be consistent with the empty lists and vice versa. I think we can all agree on that.

Let's look at how sets work in Nix then. In sets, we always put whitespace before/after the squirly braces like this `{ foo = "bar"; }` or `{ ... }`. This is consistent with multi-line sets where there is also always a whitespace (newline) before/after squirlies:

```Nix
{
  foo = "bar";
}
```

I'd find it extremely weird and inconsistent to suddenly not put whitespace before/after squirlies when it's a single line (`{foo = "bar";}`).

I've yet to see a good reason for omitting these spaces, especially given that almost all Nix code in existence has them.

Author: Tentatively implemented in https://github.com/serokell/nixfmt/pull/118/commits/c4b930210270277a658183a034e4a97374daacf0

Reviewer 17: I was mostly agnostic, but after seeing the examples in that implementation commit I'm now firmly on team spaces.

The double-winky-grinch emoticon at the end of this line is what did it for me:
```
  {a = with b; {c = 1;};}
```

Reviewer 2: Discussed again in today's meeting. For consistency reasons we really think that there should always be a space for now, both for attribute sets, lists, and empty versions of them. We'd like to resolve this topic as we don't think it's worth discussing further for now.

- Lists and attribute sets with multiple items are liberally expanded.
  - They can only be on a single line if:
    - They contain at most one element
    - Fit on the line
  - As described under [Binders](#binders) below, nested attribute sets are always expanded.

**Examples**

```nix
#1
[
  { }
  { foo = "bar"; }
  {
    foo = {
      bar = "baz";
    };
  }
  { foo.bar = "baz"; }
]

#2
[
  [ 1 ]
  [
    2
    3
  ]
]

#3
[
  [
    1
    2
    3
  ]
]

#4
[
  {
    mySingletons = [
      [
        ({
          # stuff in there
        })
      ]
    ];
  
    mySingletons' = [
      [
        (function call)
      ]
    ];
  }
]
```

**Drawbacks**

- Singleton lists may use a lot of indentation

**Rationale and alternatives**

- Have a special compact form for singleton lists, to reduce the indentation level and remove two additional lines
  ```nix
  foo = [ {
    # content
  } ];
  ```

> - Be a bit less eager about expanding lists and attribute sets (for example by allowing up to n elements on a single line).

Reviewer 12: Yes, and the following heuristic could work: ~5 non-nested elements (average human short-term memory) that fit on a line.

Reviewer 2: Maybe this decision should be moved out of the RFC. I don't think it matters whether it's 1 or 5, just needs to be some number, and we shouldn't be afraid to change it if it turns out a different one works better.

Reviewer 2: We now changed this section to allow "few" elements on the same line, without giving a specific number: https://github.com/NixOS/rfcs/pull/166/commits/385a914e9a506b6b24506a770d71f29e3cd65fdb

Reviewer 18: You call it flexible, but to me, this looks like an inconsistency. I hope not.

Reviewer 4: Without inconsistency, we either have
- cluttered diffs (when expanding lists less eagerly) or
- wasted space, eyesight clutter (if we expand lists more eagerly).

With inconsistency, the human can decide whether a specific list should be expanded or not (up to a point where the list length hits the line length limit).

However, I also really like the premise of the same code having the same output formatting, no matter the input formatting.

Reviewer 4: This happens because alejandra has no line length limit.

Reviewer 16: > kamadorueda/alejandra#373

This is not a bug, but a feature, that I would expect this RFC to provide as well. As long as the line limit allows, the formatter should respect the author's decision whether a list / attrset should remain on a single line or not. A common case when humans make better decisions that formatters is when you have two or more list/attrset bindings which are strongly related. In such cases, alignment can make the code easier for the human reader. There are various tools that automate this, like https://github.com/junegunn/vim-easy-align

> However, the output format should be as pure a function of the AST as possible, and thus as independent of the input format as possible.

Why? Are we optimizing the output for humans or for machines? If we want simplicity and consistency, we can just print everything on one line. The Nix evaluator won't mind. I hope we agree that this is not what we want. Optimizing for diff readability is optimizing for humans, so it inherentlymakes the formating non-consistent. We should embrace that, instead of fighting our nature, if we want satisfying results. Of course, consistency is important, but we should not take it to extremes (e.g. printing everything one line).

>> want to be able to write code (or to emit generated code) onto a single line with full confidence that the formatter will do its job correctly.

> This happens because alejandra has no line length limit.

Yes, indeed, it's a question of whether a particular formatter supports line limit or not. In general I find Prettier's output to be one of the best across all the formatters I've used for any programming language.
In their [rationale](https://prettier.io/docs/en/rationale.html)
they claim that they consider this a workaround and not a feature, but after having using both Prettier and Alejandra for a long time I find this a good design tradeoff, as it ultimate works very satisfyingly in practice.

Reviewer 19: > > However, the output format should be as pure a function of the AST as possible, and thus as independent of the input format as possible.
>
> Why? Are we optimizing the output for humans or for machines? If we want simplicity and consistency, we can just print everything on one line. The Nix evaluator won't mind. I hope we agree that this is not what we want. Optimizing for diff readability is optimizing for humans, so it inherentlymakes the formating non-consistent. We should embrace that, instead of fighting our nature, if we want satisfying results. Of course, consistency is important, but we should not take it to extremes (e.g. printing everything one line).

This. I 100% agree. Trying to format the code based solely on the AST is an anti-pattern and I don't really understand, why do some people insist on it. Stuff like newlines, empty lines and comments are not represented in the AST, but are extremely important for good formatting. Instead of operating on the AST, formatters should work with the CST which includes sufficient information to perfectly reconstruct arbitrary source code (including all the non-functional stuff like whitespace and comments).

Reviewer 19: I think, that it's important to recognize that humans are going to try to optimize how things look anyway. Even if we make the formatter "a pure function of the AST (+comments? +empty lines?)", that still won't prevent people from adjusting their code so that it looks better when formatted. Stuff like inserting parentheses, comments, empty lines, extracting intermediate expressions into `let` blocks, etc etc.

My reading of the `prevent future debate around how things should be formatted` goal is that we want to avoid bikeshedding about code formatting in PRs while still keeping a generally uniform style across nixpkgs. In other words, running the formatter should automatically produce code that is "good enough" to pass the PR smell check, not that the tool should completely eliminate the possibility of having two equivalent pieces of code that are formatted differently.

IMHO, within the space of "good enough to get the PR accepted" there is some space for manual adjustment by the author (who is a human and therefore may be able to format code better than just "good enough"). Rules like "keep empty lines for logical grouping", "manually decide whether to split a small list or to keep it on a single line", etc seem pretty reasonable and I don't expect them to lead to much "`future debate around how things should be formatted`". If a PR reviewer is nitpicky enough to block the PR due to such formatting choices, no amount of "pure deterministic AST formatting" is going to stop them from bikeshedding.

Author: A primary goal of this RFC is to not bother humans with the format of the code, *at all*. All stated goals of this RFC directly derive from this, although I admit that it should be written down explicitly as well.

There are two main issues with respecting formatting choices of the input code:

1. The *possibility* to make things slightly prettier will always result in an *expectation* to do so (even if it's only an internally motivated one, I know my perfectionist friends), weakening the main point of this RFC. Those attempts at tweaking the output might also lead to "fighting the auto formatter", which is frustrating and wasteful.
2. The formatter has no ability do distinguish between "I just vomited all my code onto one line because it's not my job to format it" and "I deliberately put things onto one line because I think it looks better". Therefore, preserving aspects of the input format automatically means that *everyone* has to pre-format their code, or else it will look significantly worse. This is not acceptable.

Yes humans can produce a better and prettier output than a tool. But the goal is to have the format *good enough*, and nobody should waste their time trying to make it slightly better.

---------

@RuRo
Obviously nixfmt currently uses some form of CST, or else it wouldn't be preserving comments and empty lines. I prefer think of it as an "annotated AST", but technically it matches the CST definition. Anyways, details.

I think that preserving empty lines and nothing else is a good compromise, as doing so does not exhibit most of the drawbacks I described above.

- Most code looks good enough without anyways, so nobody is forced to add them to have their code look good
- They are a crucial tool to group items together, especially given that the formatter destroys all other kinds of grouping (in a list, every item must go onto its own line, you can't do "one line per group")
- They do not really affect whether or not a syntax element will be expanded or not, because you can only really add them into already expanded code

--------

> Even if we make the formatter "a pure function of the AST (+comments? +empty lines?)", that still won't prevent people from adjusting their code so that it looks better when formatted. Stuff like inserting parentheses, comments, empty lines, extracting intermediate expressions into let blocks, etc etc.

If people are improving their code structure, that's always a win. This is also where the "Bad code does not deserve good formatting." mantra comes from: We don't want to try to make shitty code bearable (at least beyond a certain point; it's a gradient), we want to encourage people to make their code less shitty in the first place. But this is more a secondary effect than a goal on its own.

Reviewer 4: An example on how one could currently fight the formatter:
```nix
{
  foo = [
    5
    14
    6
    23
    1
    54
  ];
  bar = [5] ++ [14] ++ [6] ++ [23] ++ [1] ++ [54];
}

```

Author: I think that having formatting directives in the code could be an acceptable compromise here. They do allow customization for those who really want to, but it doesn't cause the main issue of leaving the formatter at a guess about whether a certain input format was "intentional" or not.

I'd like to note though that all of this is essentially an implementation detail based off nixfmt's current design, and that it has little to no influence over the format described in this RFC: This RFC mostly specifies the layout of individual syntax elements when expanded onto multiple lines, on most accounts it (by design) remains rather vague about the rules of when to expand a syntax element or not. I'd consider any such changes as "minor" which could thus easily be tweaked afterwards if necessary.

Reviewer 4: Adding comments can also work as a way to fight the formatter:
```nix
{
  foobar = lib.loooongfoobarfun (bar: foo: eggspam (foosomething foo) (barsomething bar)) (
    builtins.readDir ./foobar
  );
  fmtbar =
    lib.loooongfoobarfun # wrap!
      (bar: foo: eggspam (foosomething foo) (barsomething bar))
      (builtins.readDir ./foobar);
}

```

Reviewer 2: Discussed this in the meeting today. We think that some "fights" between the user and the formatter are unavoidable, but this is more of an implementation detail and doesn't need to be written down in the RFC.

Regarding the idea of allowing users to influence the formatter _just_ from the input layout, the shepherds think that it's not worth doing that compared to the current proposal as stated from the arguments in this thread.

  ```nix
  [ x y ]
  ```

> ### Binders and inherit

Reviewer 12: This is a binder: https://duckduckgo.com/?t=ffab&q=binder&iax=images&ia=images

```suggestion
### Assignments and inherit
```

Reviewer 20: First time I see _binders_.

```suggestion
### Bindings and inherit
```

Perhaps this is less specific (although I wouldn't know exactly what it refers to), but it does not create ambiguity.

Reviewer 2: Changed it to bindings! https://github.com/nix-rfc-101/rfcs/commit/9ac50f80ab4aafa6fe773a54a8549e3e01ad7fec

# If we add more attributes, the indentation stays the same, see #7
split =
  { x, ... }:
  {
    inc = x + 1;
    dec = x - 1;
  };

#7 with enough attribute bindings, they will start on a new line and become multline 
# it is common to expand input arguments in a way that does not change the number of function applications necessary
outputs =
  {
    x,
    y,
    z,
    ...
  }:
  {
    result = x + y;
  };
```

Notable special cases are:

- Single line values that would not benefit from style #2 keep using #1, even if this makes it go above the line limit. This mostly applies to simple strings and paths.
- Attribute set values are *always* expanded. This has the consequence of always forcing nested attribute sets to be multiline (even if they would be single line otherwise because they only contain a single item), which usually is desired.
  ```nix
  {
    foo.bar.baz = "qux";
    foo' = {
      bar.baz = "qux";
    };
  }
  ```
- "statement-like" expressions like "let", "if" and "assert" always use #4 (or #1).
- If the value is a `with` followed by a function application or list or attribute set, try to use #3.
  ```nix
  buildInputs = with pkgs; [
    some
    dependencies

>   ];

Reviewer 12: This doesn't have to be an exception if there's a general rule for `with` as suggested in https://github.com/NixOS/rfcs/pull/166/files#r1394820149.

  ```

**Alternatives**

One could eliminate style #2 by having #4 always start on the first line. This would even reduce indentation in some cases. However, this may look really weird in other cases, especially when the binder is very long:

```nix
some.very.long.attr = callFunction
  arg1
  arg2
  arg3;
```

#### inherit

**Description**

The items are either all on the same line, or all on a new line each (with indentation).

**Examples**

```nix
inherit foo bar baz;
inherit
  foo'
  bar'
  baz'
;
```

#### inherit from

**Description**

> - If the inherit target is single-line, it is placed on the same line as the `inherit`, even if the following items do not fit onto one line.

Reviewer 12: ```suggestion
- If the inherit source is single-line, it is placed on the same line as the `inherit`, even if the following items do not fit onto one line.
```

Not sure if that already has a term, but I'd rather call it a *source*, since that's where we inherit *from*.

  - In that case, the remaining items are force-expanded too, even if they would have fit onto one line in the first place.

**Examples**

```nix
inherit (pkgs) ap1 ap2 ap3;
inherit (pkgs)
  app1
  app2
  # …
  app42
;
inherit
  (pkgs.callPackage ./foo.nix {
    arg = "val";
  })
  attr1
  attr2
;
```

#### Semicolon placement

**Description**

The semicolon is always placed on the same line as the expression it concludes.

**Examples**

```nix
{
  attr1 = bar;
  attr2 = function call {
    # stuff
  };
  attr3 =
    function call
      many
      arguments;
  attr4 =
    let
      foo = "bar";
    in
    some statement;
  attr5 =
    if foo then
      "bar"
    else
      "baz";
  attr6 =
    let
      foo = false;
    in
    if foo then "bar" else "baz";
  attr7 = function (
    if foo then
      "bar"
    else
      "baz"
  );
  attr8 =
    cond1
    || cond2
    ||
      some function call
      && cond3;
}
```

**Rationale and alternatives**

There are four considered semicolon styles:
1. On a new line without indentation.
  - This clearly marks a separation between attributes, however it is wasteful of space.
  ```nix

>   attr3 =
>     function call
>       many
>       arguments;
>   attr3 =
>     let
>       foo = "bar";
>     in
>     some statemens;

Reviewer 12: That would match the description:

```suggestion
  attr3 =
    function call
      many
      arguments
  ;
  attr3 =
    let
      foo = "bar";
    in
    some statemens
  ;
```

Reviewer 2: Oh yeah we accidentally formatted these lines even though they're in the alternatives, fixed in https://github.com/NixOS/rfcs/pull/166/commits/63a2fa906bbf5929e9b4d82b7047619e39ea718f.

  ```
2. On a new line with one indentation level.

>   - Just as wasteful on space as (1), but a bit less clear about signaling the end of the binding.

Reviewer 12: I'd argue on the basis of low-tech ergonomics that this is preferable over (1): Editors will reasonably put you on the same level of indentation as the previous line, so if you want to terminate the sequence, just hit `;` and `Enter`. It's a reasonable trade for slight visual ambiguity, and it's not wrong to end at the same level of indentation as the expression to be assigned.

I'm strongly in favor in of the quirky-looking but highly diffable "Nix tail" at least for `inherit`. But having that for anything that's not a closing brace is probably good.

Reviewer 2: I don't think we need to optimise for typing code. Formatters are there so you can just type whatever you want and it will format it correctly.

For binders/assignments, we deemed putting the semicolon on its own line as too much deviation from the conventional style, it just looks too weird.

However for `inherit`'s we agreed on this style:
```
inherit (foo)
  bar
  baz
;
```

This section should be rewritten to be clearer about this though. I'd also be in favor of indenting the semicolon here btw, this looks better:

```nix
inherit (foo)
  bar
  baz
  ;
```

Reviewer 12: > I don't think we need to optimise for typing code. Formatters are there so you can just type whatever you want and it will format it correctly.

I'm not arguing for *optimizing* for typing, but as a tiebreaker and background guideline that's in line with this RFC's motivation: reducing friction in the contribution and review process.

Consider the case of people contributing tiny changes through the GitHub web UI: Can we make it more likely that what they type won't get stuck in CI? These will usually be beginners with the Nix ecosystem or coding in general. A red CI check is prone to scare them off, or delay a review indefinitely because they may never get to figure out how to fix it. More advanced contributors may install Git and a plain text editor, but how likely is it that they will have full development environment that happens to do all the right things, including running the formatter?

You may not see these barriers, but they are very real for first-time contributors that may never come again because of some particular papercut that got them off track.

Author: Implemented in https://github.com/serokell/nixfmt/pull/118/commits/1446d98049e50bf3cf5589adc0d16ce6de602d8c. Though this change is probably best viewed in the file view rather than as diff.

  ```nix

>   inherit (pkgs)
>     app1
>     app2
>     # …
>     app42;
> 
>   attr3 =
>     function call
>       many
>       arguments;

Reviewer 12: That would match the description:

```suggestion
  inherit (pkgs)
    app1
    app2
    # …
    app42
  ;

  attr3 =
    function call
      many
      arguments
    ;
```

  ```
3. A mix of (1) and (2), where usually the semicolon is placed directly at the end of the binder.
   But with exceptions in which the semicolon is placed onto the following line instead in cases where the value is a multiline `if` expression or nested operator.
   These are the only syntax elements which may result in the semicolon being placed on a line with arbitrarily deep indentation.
   
   ```nix

>    attr4 =
>      if foo then
>        "bar"
>      else
>        "baz";
> 
>    attr5 =
>      let
>        foo = false;
>      in
>      if foo then
>        "bar"
>      else
>        "baz";
>        
>    attr7 =
>      cond1
>      || cond2
>      ||
>        some function call
>        && cond3;

Reviewer 12: That would (probably?) match the description:

```suggestion
   attr4 =
     if foo then
       "bar"
     else
       "baz"
     ;

   attr5 =
     let
       foo = false;
     in
     if foo then
       "bar"
     else
       "baz"
     ;
       
   attr7 =
     cond1
     || cond2
     ||
       some function call
       && cond3
     ;
```

>    ```

Reviewer 12: Absolutely stellar work! ❤️❤️❤️ Thanks so much for pulling through. I fully subscribe to the motivation and the full extent of the proposal. Once implemented, it will surely eliminate a significant source of friction for contributors and maintainers.

Great piece of writing, and a beautifully painted bike shed.

---
feature: nix_formatting
start-date: 2021-08-17
author: piegamesde
co-authors: infinisil
pre-RFC reviewers: tomberek, 0x4A6F

> shepherd-team:

Author: This is more a formality than anything else, but I nominate @tomberek @0x4A6F to continue shepherding this (and also @infinisil, if that's allowed as co-author)

Reviewer 3: The RFC Steering Committee discussed this and we think it is fine to have a co-author on the shepherd team. However in this case it would be good to try to get an extra shepherd over the minimum to ensure diversity of opinions on the team. However we do no think that is strictly required.

Reviewer 2: I also accept the nomination!

Reviewer 21: accept shepherd nomination

Reviewer 22: I'd like to nominate myself as a shepherd :)

Reviewer 2: See https://github.com/NixOS/rfcs/pull/166#discussion_r1426605181 :)

Any attempt at creating an exhaustive rule set would be futile.
Therefore, the formatting rules are intentionally under-specified, leaving room for the formatter implementation.
However, the most important or potentially controversial rules are included, as well as some general meta-rules.

When deciding between two *equally good* options, currently prevalent formatting style in Nixpkgs should be followed.
The emphasis here is on "equally good".
We should not fear making radical changes to the current style if there are sufficient arguments in favor of it.

*Bad code does not deserve good formatting.*

## Detailed design

### Standard Nix format

The _standard Nix format_ defines the officially recommended way that Nix code should be formatted.

The initial version of the standard Nix format is defined in a section towards the end:

[Initial standard Nix format](#initial-standard-nix-format).

Significant changes to the standard Nix format must go through another RFC.

The latest version of the standard Nix format must be in a file on the main branch of the [official Nix formatter](#official-nix-formatter).

> ### Nix format team

Author: What would be the benefit of separating it out into two?

> ### Nix format team

Reviewer 2: Two basic ideas:
- Somebody needs to implement RFCs
- All official Nix projects need to be maintained, and not just by a single person

That's what the team is for. By putting it in the RFC, the team gets the responsibility and authority to do this job.

The _Nix format team_ is established, it has the responsibility to

> - Maintain the [official Nix formatter](#official-nix-formatter)

Reviewer 2: We have a team of 5 people from the community that are _willing_ to maintain the formatter, which I think it way better than just hoping somebody will maintain it!

> - Maintain the [official Nix formatter](#official-nix-formatter)

Reviewer 5: I like this crucial step of creating the team. 👍 Thank you so much for tackling this problem.

I still think @hugosenari has an important point here: The nix language formatting specification should be *abstractly* written down so other oss tools (e..g a lsp that warns about certain violations) can implement the same behavior without reverse engineering the 'official tool'.

> - Maintain the [official Nix formatter](#official-nix-formatter)

Reviewer 2: We mainly want to avoid formatting debates in Nixpkgs by only having one way to format Nix code, checked by CI. But it's really hard to write down a _complete_ specification of the formatting, at least from the start. By declaring a formatter as official, the formatter team can make minor adjustments for edge cases and fix bugs without having to go through another RFC every time. Everything not specified in the RFC is implicitly specified by the implementation and can be adjusted as necessary.

- Regularly [reformat Nixpkgs](#reformat-nixpkgs) with it

See the linked sections for more information.

Initially the team consists of these members:
- @piegames (author of this RFC, shepherd of the original formatting RFC)
- @infinisil (from Tweag, co-author of this RFC, shepherd of the original formatting RFC)
- @tomberek (from Flox, shepherd of the original formatting RFC)
- @0x4A6F (shepherd of the original formatting RFC)
- @Sereja313 (from Serokell, original sponsors of nixfmt)

Team member updates are left for the team itself to decide.

> ### Official Nix formatter

Author: Discussed in https://github.com/NixOS/rfcs/pull/166#discussion_r1406949589

Reformatting commits that get added to `.git-blame-ignore-revs` [won't get shown](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view) in blames on GitHub,
and can be ignored in the `git blame` command using [`--ignore-revs-file`](https://www.git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revs-fileltfilegt).

### Formatting gotchas requiring manual intervention

Some code patterns in Nixpkgs will result in a sub-optimal format,
because an auto-formatter cannot do exceptions based on context.
A lot of the times though, the same program can be equivalently expressed in a prettier way.

#### CLI argument flags

It is common to pass in CLI flags (e.g. to builders) like this:

```nix
[
  "--some-flag" "some-value"
]
```

However, this will be formatted sub-optimally:

```nix
[
  "--some-flag"
  "some-value"
]
```

If the CLI also accepts GNU-style flags, a more structured helper can be used instead:

```nix
lib.cli.toGNUCommandLine { } {
  some-flag = "some-value";
}
```

#### Singleton lists

Sometimes a list only needs a single element
and there's no expectation to add more in the future.

> This would be formatted like this:

Reviewer 23: I don't think it's reasonable to discuss that in a formatting RFC. If you use an API that expects a list, then it's alright to use a list with a single item. But to me that discussion seems clearly out of scope.

- Having a commit that changes the formatting, can make git blame harder to use.
  - This can be [worked around in interfaces](#git-blames).

## Alternatives

- Keep the status quo of not having an official formatter.
  The danger is that this creates discord within the Nix community.
  The current friction and maintainer churn due to bad formatting may arguably be small, but not negligible.
- Specify a different format.
- Specify a format, but don't enforce it in CI, to allow manually tweaking the output if necessary.
- Apply the format incrementally, i.e. only changed files get formatted, to make a more graceful transition.
  However, such an approach would not reduce the amount of merge conflicts,
  and increase the workload for contributors during that time significantly.

## Unresolved questions

- Should the line length limit be specified?

## Future work

- General style guidelines beyond AST reformatting
- Making widespread use of linters like Nixpkgs-hammering
- Enforcing the format to other official repositories

-----

## Initial standard Nix format

Terms and definitions:
- Brackets: `[]`
- Braces: `{}`
- Parentheses: `()`

> - Newlines are not guaranteed to be preserved, but empty lines are.

Reviewer 17: I think this means, but it might merit clarification if so, that the standard Nix format requires preserving multiple adjacent empty lines and does not allow collapsing them into one empty line.

Author: You are right, the implication is wrong and needs clarification (we collapse them onto single empty lines)

Reviewer 2: Now reworded and added an example for this: https://github.com/nix-rfc-101/rfcs/commit/ff2f08aa077639058711f49146e618e0dbca9057

# Bad, expressions cond and bar are fully on the same line,
# but their smallest common ancestor expression is the entire if-then-else, which spans multiple lines
if cond then foo
else bar

# Okay, cond, foo and bar have the if-then-else as a common ancestor expression,
# which is also fully on the same line
if cond then foo else bar

# Bad, due to function application precedence, the smallest common ancestor expression
# of foo and bar is `foo || bar baz`, which spans two lines
foo || bar
  baz
```

**Rationale**

This rule has turned out to be very practical at catching code that could be potentially hard to understand or edit.

### Line length

Lines have a soft length limit that doesn't count indentation.
There may also be a hard length limit that includes indentation.
String-like values such as strings, paths, comments, urls, etc. may go over the hard length limit.
These limits should be configurable, and the soft length limit should default to 100 characters.

### Indentation

- Two spaces are used for each indentation level.
  - This may be revisited should Nix get proper support for [using tabs for indentation](https://github.com/NixOS/nix/issues/7834) in the future.

> - Vertical alignment may not be persisted, neither at the start of the line nor within lines.

Reviewer 17: Does ‘may not’ here mean that complying with the standard Nix format *requires* that vertical alignment spaces be replaced with single spaces (or line breaks and indentation), or that the standard Nix format leaves it up to the formatter whether to preserve vertical alignment?

Reviewer 2: Good point, we should make sure to adhere to [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119) to be unambiguous.

Reviewer 12: Vertical alignment produces diff noise. Potentially a lot.

Reviewer 14: Regarding diff noise: we don't want to have our formatter provide "view" and "diff" mode, right? Because we can configure git to use a specific diff format even if the files are formatted differently. The goal is to explicitly **not** have that, yes?

Author: A view mode was discussed, but no conclusion was reached IIRC. Personally I wouldn't generally rule it out, but say that it's out of scope for this RFC. So things need to work sufficiently well without for now.

Reviewer: Maybe should be the case to add _[disable tags (comments)](https://astyle.sourceforge.net/astyle.html#_Disable_Formatting)_, do we plan to have any?

Subjective comment, I think nothing add more value to ~readability~ navigation of the code than vertical alignment (indentation included),

Not so subjective comment, I found data saying it has no difference.  [A Systematic Literature Review on the Impact of Formatting Elements on Code Legibility](https://arxiv.org/abs/2208.12141) (page 24, table 6)

Reviewer 2: I think such formatting tags should be out of scope for this RFC.

@hugosenari: Nice find!

Reviewer 2: Discussed this today with the shepherd team. We still don't think that there should be vertical alignment, though we generally think it would be okay to change the behavior with special formatting directives, though that is out of scope for this RFC.

> ---

Reviewer 2: Imo a Nix code formatter should never change the AST or the evaluation result (ignoring impurities like `builtins.unsafeGetAttrPos`). If at all, it could be an opt-in feature in the future. I think this should definitely be out of scope of this RFC.

> ---

Author: Currently, any strings containing bash are Nixpkgs specific and thus out of scope of this RFC. A first step towards that direction would be to have syntactic annotation of the language or "format" of the string's content, so that the formatter can work with it in the first place. But then, the other restrictions still apply.

## Future work

- General style guidelines beyond AST reformatting
- Making widespread use of linters like Nixpkgs-hammering
- Enforcing the format to other official repositories

-----

## Initial standard Nix format

Terms and definitions:
- Brackets: `[]`
- Braces: `{}`
- Parentheses: `()`

- Newlines are not guaranteed to be preserved, but empty lines are.
  - This allows the formatter to compact down multi-line expressions if necessary, while still allowing to structure the code appropriately.
- Expressions of the same kind that can be treated as a sequence of expressions on the same level should be treated as such, even though they are technically parsed as a nested tree.
  - This applies to else-if chains, functions with multiple arguments, some operators, etc.
  - Example:
    ```nix
    # This is treated as a sequence of if-then-elsa's chains
    if cond1 then
      foo
    else if cond2 then
      bar
    else
      baz
    ```

### Single-line common ancestor expression rule

For any two (sub-)expressions that are fully on a common single line, their smallest common ancestor expression must also be on the same line.

**Example**

```nix

> # Bad, expressions cond and bar are fully on the same line,

Reviewer 21: ```suggestion
# Bad, expressions cond and foor are fully on the same line,
```
typo?

Reviewer 2: Please PR for trivial suggestions to avoid clutter! And you can even push to the repo directly as a shepherd ;)

# but their smallest common ancestor expression is the entire if-then-else, which spans multiple lines
if cond then foo
else bar

# Okay, cond, foo and bar have the if-then-else as a common ancestor expression,
# which is also fully on the same line

> if cond then foo else bar

Reviewer 24: > It doesn't violate the SLCAE rule (that name's definitely going to catch on!)

Sorry for being noob, but what's that acronym stands for?

If the order isn't important, would we make it SCALE or some other pronounceable permutation?

> if cond then foo else bar

Reviewer 24: > Single-Line Common Ancestor Expression\nhttps://github.com/NixOS/rfcs/blob/299c16f9c7ff96e28c5314bc9de0b1e72656d795/rfcs/0166-nix-formatting.md?plain=1#L278-L280

Thank you!

  - The cutoff is usually determined empirically based on common usage patterns.

**Examples:**

```nix
{
  #1
  buildInputs = [
    foo
    bar
    baz

    somethingElse
  ];

  #2
  systemd.services = {
    foo = { };
    bar = { };
  };

  #3
  inherit
    lib
    foo
    bar
    baz
    ;
}
```

### Function application

**Description:**

- In a function application chain, the first element is treated as the "function" and the remaining ones as "arguments".
- The last argument receives special treatment, to better represent common coding patterns.
- As many arguments as possible are fit onto the first line.
  - If all but the last argument do fit, then the last argument may start on the same line.
  - If an earlier argument does not fit onto the first line, then itself and all the following ones start on a new line. This is called the expanded form.
  - All arguments that are not on the same line as the function are indented by one level.

**Examples:**

```nix
#1
function arg1 arg2

#2
function arg1 arg2 {
  more = "things";
}

#3
function arg1 arg2 arg3 # reached line limit here
  arg4
  arg5
  [
    1
    2
  ]

>   arg7

Author: This is a contrived example, yes, but in practice I think the current style is preferable most of the time. Especially since it is a common pattern to have `function short args (big lambda on new line) attrs` (e.g. `concatMapStringsSep`)

#4 All arguments that don't fit on the same line, start indented on a new line
function arg1
  {
    # items that don't fit on same line
  }
  arg3

#5
function
  {
    a = 1;
    b = 2;
  }
  {
    c = 1;
    d = 2;
  }

#6 (with "--width=20")
function arg1 (
  function2 args
)
```

**Drawbacks**

- This style sometimes forces lists or attribute sets to start on a new line, with additional indentation of their items.

**Rationale and alternatives**

> - Not indenting the arguments, to save some indentation depth. This would be consistent with other constructs like function declarations and let bindings.

Reviewer 5: It might be consistent to either indent all arguments.

```nix
func1 
  arg1
  func2
  arg2
```

Or maybe I misinterpreted your example?

> - Not indenting the arguments, to save some indentation depth. This would be consistent with other constructs like function declarations and let bindings.

Author: Indenting all arguments is the current style, this is just the discussion of an alternative which was previously explored

> - Not indenting the arguments, to save some indentation depth. This would be consistent with other constructs like function declarations and let bindings.

Reviewer 2: We should maybe put the alternatives in a `<details>` to hide it by default

> - Not indenting the arguments, to save some indentation depth. This would be consistent with other constructs like function declarations and let bindings.

Author: Would this make a difference for the review view?

> - Not indenting the arguments, to save some indentation depth. This would be consistent with other constructs like function declarations and let bindings.

Reviewer 14: I noticed that it is in "alternatives" block. My example is pretty bad, so here's a better one, taken from example number 3.

```
#3 with indent (as currently proposed)
function arg1 arg2 arg3 # reached line limit here
  arg4
  arg5
  [
    1
    2
  ]
  arg7

#3 no indent (what this alternative suggests, unless I'm mistaken)
function arg1 arg2 arg3 # reached line limit here
arg4
arg5
[
  1
  2
]
arg7
```

What I'm trying to say is that example number 3 without indentation at all looks **awful**. I personally wouldn't even consider it, that's how ugly it looks.

> - Not indenting the arguments, to save some indentation depth. This would be consistent with other constructs like function declarations and let bindings.

Reviewer 2: Now removed as an alternative in https://github.com/nix-rfc-101/rfcs/commit/e94e1e3d4c3368c4c4aead1c9fde7a6e1c53197f :)

    with bar;
    [
      # multiline
      baz
    ]

  # Good
  foo = with bar; [
    # multiline
    baz
  ];
  
  # Good
  foo =
    with bar;
    baz foo {
      # multiline
      qux = 10;
    }
    
  # Good
  foo = with bar; (
    baz
  );

  # Good
  foo =
    with bar;
    if cond then
      foo
    else
      bar;
  
  # Bad
  foo = assert qux; with bar; [
    # multiline
    baz
  ];
  
  # Bad
  foo = with bar;
    [
      # multiline
      baz
    ];

  # Bad
  foo =
    with bar; [
      # multiline
      baz
    ];
}
```

**Examples**

```nix
with pkgs;
assert foo == bar;
{
  meta.maintainers = with lib.maintainers; [
    some
    people
  ];
}

with pkgs;
assert foo == bar;
{
  meta.maintainers =
    with lib.maintainers;
    [
      some
      people
    ];
}
```

### let

**Description**

- Let bindings are *always* multiline.
- The "let" part is indented one level, but not the "in" part.
- Each item in the "let" part is indented and starts on its own line.
  For more details, see the [bindings section](#bindings).
- The "in" part starts on a new line.

**Examples**

```nix
let
  foo = "bar";
in
if foo == "bar" then
  "hello"
else
  "world"

> ```

Reviewer 2: Now added your examples first in https://github.com/nix-rfc-101/rfcs/commit/e94e1e3d4c3368c4c4aead1c9fde7a6e1c53197f :)

Any release of the official Nix formatter must conform to the latest version of the [standard Nix format](#standard-nix-format).

The latest release of the official Nix formatter should support the Nix language syntax of the latest Nix release.
The Nix format team should be consulted before the Nix language syntax is changed.

### Reformat Nixpkgs

For formatting Nixpkgs itself, a pinned release version of the official Nix formatter must be used.
CI must generally enforce all files in Nixpkgs to be formatted with this version at all times.
Automatically generated files may be handled differently, see [the next section](#automatically-generated-files).

For every bump of the pinned formatter,
the files of Nixpkgs must thus be re-formatted accordingly.
Commits that reformat Nixpkgs will be added to `.git-blame-ignore-revs`,
which can then be [ignored in tooling](#git-blames).
The Nix format team is responsible for this task.

In order to minimize conflicts especially when back-porting,
the pinned formatter should preferably only be updated shortly before the release branch-off.

> This should be done in coordination with the NixOS release managers.

Reviewer 22: This should be added to the release wiki to make sure it's never forgotten ;)

Reviewer 2: Added in https://github.com/nix-rfc-101/rfcs/commit/e94e1e3d4c3368c4c4aead1c9fde7a6e1c53197f :)

- Not indenting the arguments, to save some indentation depth. This would be consistent with other constructs like function declarations and let bindings.
- Compacting multiline arguments like this:
  ```nix
  #4b
  function arg1 {
    # stuff
  } arg3

  #5b
  function {
    # …
  } {
    # …
  }
  ```
  - This violates the guideline of the indentation representing the expression structure, and thus reduces readability.
  - This does not work well with line length limits on short arguments like in example #3.

### Function declaration

**Description**

- The body of the function is not indented relative to its arguments.
- Multiple ("simple") identifier arguments are written onto the same line if possible.
- Attribute set arguments always start on a new line; they are not mixed with identifier arguments.
  - If they have few attributes, the argument may be written on a single line, otherwise the expanded form is used.
- Attribute set arguments have their attributes on a new line each with indentation, followed by a trailing comma.

**Examples**

```nix
#1
name: value: name ++ value

#2
name: value:
{
  "${name}-foo" = value;
}

#3
{ pkgs }: pkgs.hello

#4

> args@{

Reviewer 22: Does `args@` always come at the front? I'm asking because I usually do `{ whatever }@args` which is also valid syntax.

Reviewer 2: Because both is allowed in the Nix syntax, the formatter must also support both. I guess this should be a general rule written down somewhere.

**Example**

```nix
# Bad, expressions cond and bar are fully on the same line,
# but their smallest common ancestor expression is the entire if-then-else, which spans multiple lines
if cond then foo
else bar

# Okay, cond, foo and bar have the if-then-else as a common ancestor expression,
# which is also fully on the same line
if cond then foo else bar

# Bad, due to function application precedence, the smallest common ancestor expression
# of foo and bar is `foo || bar baz`, which spans two lines
foo || bar
  baz
```

**Rationale**

This rule has turned out to be very practical at catching code that could be potentially hard to understand or edit.

### Line length

- There should be a configurable _soft_ line length limit, limiting the number of characters on one line without counting the leading indentation.
  The default should be 100 characters.
- There may also be a configurable _hard_ line length limit, which includes the leading indentation.
- String-like values such as strings, paths, comments, urls, etc. may go over the hard line length limit.

### Indentation

> - Two spaces must be used for each indentation level.

Reviewer 8: @toastal I definitely have the same issue. In my personal Nix files repo I set a wider indentation (with tabs) because it was too difficult to keep track of indentation otherwise. Often when reading Nixpkgs code I have to reformat the indentation or add comments that note different ends of indentation blocks to be able to understand what I'm looking at. It's definitely not very accessible to me, at least.

> - Two spaces must be used for each indentation level.

Reviewer 25: I'm curious if structure-guide lines help you at all.

> - Two spaces must be used for each indentation level.

Reviewer 8: Surprisingly not really. It can help a little in some cases but it also adds more visual noise.

> - Two spaces must be used for each indentation level.

Reviewer 4: @toastal if you use vscode or vscodium, you can use this extension to fix it on your end: https://marketplace.visualstudio.com/items?itemName=kylepaulsen.stretchy-spaces

i have not found anything similar for vim yet tho

> - Two spaces must be used for each indentation level.

Reviewer 2: While these are good arguments, 2 spaces of indentation is by far the most popular style right now, and we consider changing this out of scope for this RFC. Furthermore, the next sentence in the RFC states:

> This may be revisited should Nix get proper support for [using tabs for indentation](https://github.com/NixOS/nix/issues/7834) in the future.

> - Two spaces must be used for each indentation level.

Reviewer 2: Currently it's de-facto standard to use 2 spaces of indentation, that's what everybody has specified in their editor configs and has been using for the past 20 years. Changing this would be a substantial change on its own and deserve its own RFC. I even think it would be deceiving to now include this in the RFC after it has gone all this time without even mentioning this. So for anybody that's passionate about this, please open a separate RFC!

Generally, to make an RFC succeed, it should be focused and as minimal as possible. So really, we shouldn't even mention how indentation is done at all, it's really not very relevant for this RFC.

- Having a commit that changes the formatting, can make git blame harder to use.
  - This can be [worked around in interfaces](#git-blames).

## Alternatives

- Keep the status quo of not having an official formatter.
  The danger is that this creates discord within the Nix community.
  The current friction and maintainer churn due to bad formatting may arguably be small, but not negligible.
- Specify a different format.
- Specify a format, but don't enforce it in CI, to allow manually tweaking the output if necessary.
- Apply the format incrementally, i.e. only changed files get formatted, to make a more graceful transition.
  However, such an approach would not reduce the amount of merge conflicts,
  and increase the workload for contributors during that time significantly.

## Unresolved questions

- Should the line length limit be specified?

## Future work

- General style guidelines beyond AST reformatting
- Making widespread use of linters like Nixpkgs-hammering
- Enforcing the format to other official repositories

-----

## Initial standard Nix format

Terms and definitions:
- Brackets: `[]`
- Braces: `{}`
- Parentheses: `()`

> - Line breaks may not be preserved, but empty lines must always be preserved.

Reviewer 17: More ‘may not’ pedantry:

```suggestion
- Line breaks may be added or removed, but empty lines must not be added, and empty lines must be removed if they follow another empty line (and must not be removed otherwise).
```

Also, while I'm picking this nit: would we want this to say that exactly one line break (i.e., no empty lines) must appear at the end of a file?

Reviewer 26: EOL on EOF.

Reviewer 2: > Also, while I'm picking this nit: would we want this to say that exactly one line break (i.e., no empty lines) must appear at the end of a file?

We just added an editorconfig yesterday:
https://github.com/nix-rfc-101/rfcs/commit/5fd936505c25533875cb23ae32ce69a018516e09

Reviewer 2: @rhendric Regarding the wording on line breaks and various other improvements, see https://github.com/nix-rfc-101/rfcs/pull/15. In particular I'm changing the wording to this there:

> Line breaks may be added or removed. Consecutive empty lines must be turned into a single empty line.

---
feature: nix_formatting
start-date: 2021-08-17
author: piegamesde
co-authors: infinisil
pre-RFC reviewers: tomberek, 0x4A6F

> shepherd-team:
> shepherd-leader:

Reviewer 27: ```suggestion
shepherd-team: das_j, tomberek, 0x4A6F, infinisil
shepherd-leader: infinisil
```

- No automatic code format can be as pretty as a carefully crafted manual one. There will always be "ugly" edge cases.
  - However, we argue that on average the new format will be an improvement,
    and that contributors should not waste their time making the code slightly more pretty.
- Every formatter will have bugs, but the Nix formatter team will be able to make fixes.
- Having a commit that changes the formatting, can make git blame harder to use.
  - This can be [worked around in interfaces](#git-blames).

## Alternatives

- Keep the status quo of not having an official formatter.
  The danger is that this creates discord within the Nix community.
  The current friction and maintainer churn due to bad formatting may arguably be small, but not negligible.
- Specify a different format.
- Specify a format, but don't enforce it in CI, to allow manually tweaking the output if necessary.
- Apply the format incrementally, i.e. only changed files get formatted, to make a more graceful transition.
  However, such an approach would not reduce the amount of merge conflicts,
  and increase the workload for contributors during that time significantly.

## Unresolved questions

> - Should the line length limit be specified?

Reviewer 10: I was just experimenting with the current implementation and realized that if I change the line length limit to something small, format my code, then change it back to default, some of the now multi-line expression get brought back to one-liners. If this is going to be the final behavior, I don't think line length should be configurable, as it no longer creates a standard, but multiple incompatible standards. If however, formats from smaller line length limits are unchanged by larger line length limits, I think keeping it configurable is probably okay.

Reviewer 10: I suppose another solution could be to keep the current behavior, and make it configurable on a per file basis via a comment in the file, but I can't say I'm a huge fan of that.

Author: I won't go into whether or not the formatter should do this, as it has been extensively discussed in https://github.com/NixOS/rfcs/pull/166#discussion_r1394844307 already.

About the question of the implications for the configurable line length limit however:

I think that it is fine as-is. The format detailed in the RFC only focuses on the format of the expanded (multi-line) form of expressions, together with "Unless stated otherwise, any expression that fits onto one single line will be trivially formatted as such".

This deliberately leaves a lot of freeroom to define what means to "fit onto one single line" in many but not all cases, and thus allows to configure the line length limit. I'd disagree with your "multiple incompatible standards",  and instead of think of it as different "flavors" of the same standard.

Reviewer 10: The reason I do not see it as different flavors of the same standard is this: nixpkgs is going to have to pick a line length setting. If nixpkgs is going to be linted based on this standard, developers will not be able to submit code that is formatted with any other "flavor". And If a particular formatting doesn't qualify for use in nixpkgs, I have a hard time seeing it as "standard".

Author: Yes, there will be one "blessed" configuration used for Nixpkgs itself, and it will be pinned in the repo with tooling for everyone to automatically use it. This doesn't mean other Nix related projects shouldn't be able to set a different line length limit if desired.

Reviewer 2: Furthermore, third-party projects probably don't want to use a vastly different line length limit than the default one, so submitting to Nixpkgs will require few changes, while the overall style and feeling of the format is the same.

We think this is fine and not a big problem overall.

- If the value is a `with` followed by a function application or list or attribute set, try to use #3.
  ```nix
  buildInputs = with pkgs; [
    some
    dependencies
  ];
  ```

**Alternatives**

One could eliminate style #2 by having #4 always start on the first line. This would even reduce indentation in some cases. However, this may look really weird in other cases, especially when the binding is very long:

```nix
some.very.long.attr = callFunction
  arg1
  arg2
  arg3;
```

#### Bindings semicolon placement

**Description**

The semicolon is always placed on the same line as the expression it concludes.

**Examples**

```nix
{
  attr1 = bar;
  attr2 = function call {
    # stuff
  };
  attr3 =
    function call
      many
      arguments;
  attr4 =
    let
      foo = "bar";
    in
    some statement;
  attr5 =
    if foo then
      "bar"
    else
      "baz";
  attr6 =
    let
      foo = false;
    in
    if foo then "bar" else "baz";
  attr7 = function (
    if foo then
      "bar"
    else
      "baz"
  );
  attr8 =
    cond1
    || cond2
    ||
      some function call
      && cond3;
}
```

**Rationale and alternatives**

> There are four considered semicolon styles:

Reviewer 25: Seems like there are only three styles here. The "four" must be including the style selected above but that could be more clear. 
```suggestion
There are three considered alternative semicolon styles:
```

Team member updates are left for the team itself to decide.

### Official Nix formatter

The Nix formatter team is given the authority and responsibility of
creating and maintaining the _official Nix formatter_ implementation.
This is a repository in the NixOS GitHub organisation.
The repository will initially be based on [this nixfmt pull request](https://github.com/serokell/nixfmt/pull/118).
This pull request has been developed along with this RFC and is already reasonably close to the proposed initial standard format.

Any release of the official Nix formatter must conform to the latest version of the [standard Nix format](#standard-nix-format).

The latest release of the official Nix formatter should support the Nix language syntax of the latest Nix release.
The Nix formatter team should be consulted before the Nix language syntax is changed.

### Reformat Nixpkgs

For formatting Nixpkgs itself, a pinned release version of the official Nix formatter must be used.
CI must generally enforce all files in Nixpkgs to be formatted with this version at all times.
Automatically generated files may be handled differently, see [the next section](#automatically-generated-files).

> 

Reviewer 28: ```suggestion
The Nixpkgs manual must also adhere to this Nix formatting standard and be formatted accordingly.

```

Just mentioning this, so it won't be forgotten when implementing.

Reviewer 2: Put this in its own section: 4f99905ed2b4af7360afa4cd4046c74a68396afa

- To allow having the `<body>` be on the same line as the `in`:
  ```
  let
    <name1> = <value1>;
    <name2> = <value2>;
    ...
  in <body>
  ```

  In particular when `<body>` is an identifier, list, attribute set and/or others.

  Problems with this alternative:
  - It leads to larger diffs when inserting something after the `in`
  - The formatting can change when `<body>` is updated
  - It's less consistent, since the formatting depends on the `<body>`

- The body could be indented by a level
  ```
  let
    <name1> = <value1>;
    <name2> = <value2>;
    ...
  in
    <body>
  ```
  
  Problems with this alternative:
  - Leads to indentation creeps
  - Inconsistent with other expressions that have a `<body>` that is "returned"
  - Favors a style where the body starts on the same line as the in for some values (e.g. attribute sets) to reduce an indentation level, see above.

### Attribute sets and lists

- Brackets and braces must always have a space (or line break) on the inside, like `[ `, ` ]`, `{ ` and ` }`.
  - Empty lists and attribute sets are written as `[ ]` and `{ }`, respectively.
- Lists and attribute sets with multiple items should be liberally expanded.

>   - They can only be on a single line if they fit on the line and contain few enough elements.

Author: Once you go through all relevant scenarios and edge cases, the rules specifying this may be pretty complex. It is not beneficial to nail them down in the RFC text with every detail. Instead, the text merely captures the *intent*, while leaving the actual implementation, well, as an implementation detail. As an example for edge cases, we currently have rules in Nixfmt that force-expand attribute sets in certain positions.

Generally, I like the format to mostly specify how syntax nodes look like in their expanded form (and their unexpanded form, but which is almost always trivial). On the other hand, the rules about *when* to expand something or not on are much less clearly defined.

The idea is that this results in a format that is spread-out by default but compact where it matters.

The interactions between different language features are complex and producing a style that matches the expectations involves a lot of special cases.
Any attempt at creating an exhaustive rule set would be futile.
Therefore, the formatting rules are intentionally under-specified, leaving room for the formatter implementation.
However, the most important or potentially controversial rules are included, as well as some general meta-rules.

When deciding between two *equally good* options, currently prevalent formatting style in Nixpkgs should be followed.
The emphasis here is on "equally good".
We should not fear making radical changes to the current style if there are sufficient arguments in favor of it.

*Bad code does not deserve good formatting.*

## Detailed design

### Standard Nix format

The _standard Nix format_ defines the officially recommended way that Nix code should be formatted.

The initial version of the standard Nix format is defined in a section towards the end:

[Initial standard Nix format](#initial-standard-nix-format).

> Significant changes to the standard Nix format must go through another RFC.

Reviewer 29: What constitutes a _significant_ change?

Personally, I'd defer *all* future changes to the team, but if we want to require RFCs for larger changes it may be helpful to try and define significant here.

Reviewer 2: From today's meeting we're updating this to

> Controversial changes to the standard Nix format must go through another RFC.
> It is up to the Nix formatter team to decide when an RFC is necessary.

    and that contributors should not waste their time making the code slightly more pretty.
- Every formatter will have bugs, but the Nix formatter team will be able to make fixes.
- Having a commit that changes the formatting, can make git blame harder to use.
  - This can be [worked around in interfaces](#git-blames).

## Alternatives

- Keep the status quo of not having an official formatter.
  The danger is that this creates discord within the Nix community.
  The current friction and maintainer churn due to bad formatting may arguably be small, but not negligible.
- Specify a different format.
- Specify a format, but don't enforce it in CI, to allow manually tweaking the output if necessary.
- Apply the format incrementally, i.e. only changed files get formatted, to make a more graceful transition.
  However, such an approach would not reduce the amount of merge conflicts,
  and increase the workload for contributors during that time significantly.

## Unresolved questions

## Future work

- General style guidelines beyond AST reformatting
- Making widespread use of linters like Nixpkgs-hammering
- Enforcing the format to other official repositories

-----

> ## Initial standard Nix format

Reviewer 6: I didn't know where else to open a thread for this because it is missing: Comments.

Apparently, the WIP formatter turns multi-line comments into single-line comments. This was confirmed to me as intended by @piegamesde.

I'd like for formatting rules regarding multi-line comments aswell as single-line comments to be added.

These should also include rules for where to place comments relative to the expression it refers to. I'd like to forbid placing the comment in the lines after the expression for example. This is also relevant for the auto-formatter because it must keep existing comments in sensible places too.

Reviewer 2: We just added a section to specify that the formatting of strings shouldn't be changed (https://github.com/NixOS/rfcs/pull/166/commits/8cb46fd96dc650b81b28f0d28f08805458d27951), and I'd be in favor of also doing the same for comments.

Though there could be some special cases where multiline comments aren't necessary anymore due to a reformatting, where it might make sense to turn them into single-line comments.

Reviewer 29: I agree that string format should stay as is. More than a handful of times I’ve emptied a string momentarily only for it to be converted.

Author: > Apparently, the WIP formatter turns multi-line comments into single-line comments. This was confirmed to me as intended by @piegamesde.

Just to be clear, this only applies to *some* comments. Specifically, to multi-line comments where all intermittent lines start with `*`. I don't know why this rule was added, however a quick test shows that it only affects a couple dozen of files in Nixpkgs anyways: https://github.com/piegamesde/nixpkgs/pull/4/commits/e371bdbbd11ba31976f5e558ffd9d395f3f631e2. (My hypothesis is that adding normalization was easier than getting them to render correctly in all cases)

Nevertheless, both the implementation and the RFC needs some polish here. I propose at least the following rules:

- `/**` comments must be handled according to RFC #145
- Single-line `/*` comments must be converted to `#` comments
- Single-line comments may be moved up or down a line to improve the layout

I think having some more rules which allow some kinds of content normalization, especially to keep things properly aligned, would be interesting to have. But I'd have to gather more examples on that before proposing anything.

Reviewer 6: @piegamesde please also note that there can also be the case where there is a comment *between* two expressions.

It's very common to "declare" dual-licenses like this:

https://github.com/NixOS/nixpkgs/blob/5ef42fcd84b2baa16f43554f1c1f1d614e23ef9a/pkgs/servers/static-web-server/default.nix#L38

What I also sometimes like to do this little trick to add comments to multi-line strings without adding the comment inside the string:

https://github.com/NixOS/nixpkgs/blob/5ef42fcd84b2baa16f43554f1c1f1d614e23ef9a/pkgs/development/libraries/dbus/default.nix#L38

(Though sadly I don't think this will be possible anymore.)

Reviewer 30: Could you please elaborate on the intended mental parsing of the latter example? For someone who's never touched that part of the codebase/are unfamiliar with it?

I couldn't figure out for myself what that comment tried to tell me.

Reviewer 6: @x10an14 it concatenates two multi-line strings. These two are the same value (`"foo\nbar\n"`):

```Nix
''
  foo
  bar
''
```

```Nix
''
  foo
'' + ''
  bar
''
```

But because you're outside the string declaration at the `+`, you can add an in-line comment without changing the value:

```Nix
''
  foo
'' + /* you could write anything here */ ''
  bar
''
```

Once you recognise this pattern, it becomes very easy to read and doesn't break the flow of the string. It's actually quite useful for adding comments to code that otherwise wouldn't allow a comment at such a place as is often the case with bash scripts.

The current style turns it into this:

```Nix
''
  foo
''
# you could write anything here
+ ''
  bar
''
```

There's now 3 lines of code between `foo` and `bar` eventhough there's still only a single newline in the actual string.

You could do

```Nix
''
  foo
'' # you could write anything here
+ ''
  bar
''
```

and that's permitted under the new style but the comment would actually be "connected" to the `foo` string and there's still an additional newline.

Author: We could discuss the necessity of that specific comment, and we could also discuss that in this case the comment could equally well be written as part of the string.

But the general pattern shown in this example is real and valid, since it generalized to other cases where such workarounds are not possible.

Have another arbitrary example from Nixpkgs:

```nix
      (with all;
        # disable default openssl extension
        (lib.filter (e: e.pname != "php-openssl") enabled)
        # use OpenSSL 1.1 for RC4 Nextcloud encryption if user
        # has acknowledged the brokenness of the ciphers (RC4).
        # TODO: remove when https://github.com/nextcloud/server/issues/32003 is fixed.
        ++ (if cfg.enableBrokenCiphersForSSE then [ cfg.phpPackage.extensions.openssl-legacy ] else [ cfg.phpPackage.extensions.openssl ])
        ++ optional cfg.enableImagemagick imagick
        # Optionally enabled depending on caching settings
        ++ optional cfg.caching.apcu apcu
        ++ optional cfg.caching.redis redis
        ++ optional cfg.caching.memcached memcached
      )
```

------

@Atemu the issues I have with your proposal are:

- It does not scale well to very long comments or multi-line comments. Comments are never broken up and therefore not part of any line limit calculations, therefore having any tokens (`''`, `[`) after a comment might end up placing them way off-screen.
- It only works for strings, lists and attrsets and does not generalize to other situations. As soon as functions like `lib.optional` are introduced –which is a very common pattern–, having the inline comment would quickly get weird results. The currently proposed handling of comments generalizes well to lots of situations, increasing overall consistency.

  - If an earlier argument does not fit onto the first line, then that argument and all the following ones must start on their own line.
  - All arguments that are not on the same line as the function must be indented by one level.

**Examples:**

```nix
# All arguments fit onto the first line
function arg1 arg2

# The line length limit is reached, so the remaining arguments need to be on their own lines
function arg1 arg2 arg3
  arg4
  arg5

# The last argument is a multiline expression, so it doesn't fit on the first line,
# but it can still start on the first line
function arg1 arg2 {
  more = "things";
}

# The second argument doesn't fit on the first line, but it's not the last argument,
# so it needs to start on a new line
function arg1 arg2
  {
    more = "things";
  }
  arg3

# Same with more multiline arguments
function
  {
    a = 1;
    b = 2;
  }
  {
    c = 1;
    d = 2;
  }
```

**Drawbacks**

- This style sometimes forces lists or attribute sets to start on a new line, with additional indentation of their items.

**Alternatives**

- Compacting multiline arguments like this:
  ```nix
  function arg1 {
    # stuff
  } arg3

  function {
    # …
  } {
    # …
  }
  ```

>   - This violates the guideline of the indentation representing the expression structure, and thus reduces readability.

Reviewer 31: I think it would be nice if the first form was allowed.

I think being strict about this produces an unfortunate amount of indentation for no real readability gain.

```nix
    pkgsStatic.callPackage
      (
        {
          execline,
          kmod,
          mdevd,
          cryptsetup,
          util-linuxMinimal,
        }:

        # ...
      )
      { }
```

Could instead be

```nix
    pkgsStatic.callPackage (
      {
        execline,
        kmod,
        mdevd,
        cryptsetup,
        util-linuxMinimal,
      }:

      # ...
    ) { }
```

I think the structure of the expression is still perfectly clear and represented by the indentation in this case.

Reviewer 31: One practical way of achieving this could be to allow a _single_ argument to be expanded over multiple lines, but other rules to achieve the same result would also work.

Author: What you propose would be acceptable to me *as long as the last argument is single-line*. In fact, we already have a similar special case for bindings with `//` operators.

Current workarounds to this which is already used in nixpkgs is to use `flip` to put the largest argument last (although I'm personally not that fond of it). Once this RFC is merged I hope continue working on the pipe operator RFC, because many of the cases with ugly formatting could be written in a cleaner way using pipes.

Reviewer 31: > What you propose would be acceptable to me _as long as the last argument is single-line_. In fact, we already have a similar special case for bindings with `//` operators.

That would be perfectly fine for me.

Author: I've pushed a first draft implementation to the nixfmt branch, feel free to try it out. I think the simplification you proposed is an improvement in many cases, however I still need to tweak the rules to sort out some false positives where it shouldn't be applied.

Reviewer 31: Oh, sorry, I misread your proposal.  "as long as the last argument is single-line" doesn't work for callPackage, because you usually end with a `{ }` after your big package definition function.  So this doesn't actually solve the example I proposed.

Reviewer 31: (The file I'm formatting here is https://spectrum-os.org/git/spectrum/tree/host/initramfs/default.nix?id=1a58ed4c6b9b5fbfb3f8d2aad36cf84dfb77ac58 if it's helpful to reproduce)

Author: I ditched the prototype implementation and did lots of refactoring to implement this properly. It should also handle your examples correctly now. Nixpkgs diff: https://github.com/piegamesde/nixpkgs/pull/4/commits/7a7a7b5036840cff8e2f7f09032180ffa1d83a79

Reviewer 32: This is _so_ much tighter without losing any readability. Very nice!

Reviewer 30: I concour! +1 to this change! =)

Reviewer 4: I really liked the way function applications with complex arguments were formatted before. It created a clear visual separation between the arguments.

Author: I agree that while the change is mostly positive, it sometimes reduces readability. I will try to find some heuristic for cases where this should better not be applied. So if you see any changes you perceive as regressions, feel free to point them out here for sample size.

> ### Function declaration

Author: Yes, parenthesized expression in function applications and bindings are treated the same, in that the body always starts on a new line.

But I did a quick experiment: https://github.com/serokell/nixfmt/pull/118/commits/187704ca59b7b007e8414d152122325da96eb1ed, https://github.com/piegamesde/nixpkgs/pull/4/commits/68ae0750025618e20f8ecb932d9da39301143955

I tried to find some minimal rules to special-case, and also took the opportunity to force-expand attribute sets in these cases. This should be in line with the typical code in these situations, where more items are expected to be added in the future.

IMO much of the diff is an improvement, however it sometimes results in things being a bit too compact and crowded. I'll try to tone things down again a bit, probably around nested function calls, in interpolations or with string bodies.

> ### Function declaration

Reviewer 2: We now adjusted the RFC to take this change into account: https://github.com/NixOS/rfcs/pull/166/commits/c617b9f26d7c9f6201e04d91cf2ad0f47427d854

> - The body of the function must not be indented relative to its first arguments.

Reviewer 31: I'd like to suggest that, additionally, a first level of indentation should not be mandatory when a function declaration is wrapped in a set of parentheses to be passed as an argument to another function.

Consider this common pattern in NixOS tests:

```nix
import ./make-test-python.nix (
{ pkgs, lib, ... }:

let
  # ...
in

{
  # ...
}
)
```

Which would be reformatted as:

```nix
import ./make-test-python.nix (
  { pkgs, lib, ... }:

  let
    # ...
  in

  {
    # ...
  }
)
```

The function definition is effectively at the top level, so I think it's less readable to have the whole rest of the contents of the expression indented like this.

Reviewer 31: Essentially, for me, it's important for readability that boilerplate doesn't introduce the first level of indentation.  Indentation is a signal to me that I should pay attention, and the point of boilerplate helpers like make-test-python.nix is that I don't need to pay too much attention to them.  The only reason they're required at all is to make files directly buildable from the Nix CLI without having to be imported/callPackaged/whatever.  It's a quirk of the language that these are function calls at all.

Reviewer 2: We discussed this in the meeting and don't think that we should change the formatting for this. The inconsistency it introduces doesn't seem worth the reduction of one level of indentation. Our suggestion is to refactor the code to not need the boilerplate in the first place, e.g. by creating a `default.nix` that calls `make-test-python.nix` instead. And that's also what Nixpkgs' NixOS tests now do, allowing you to also call the tests with `nix-build -A nixosTests.<name>`.

- Multiple ("simple") identifier arguments must be written onto the same line if admitted from on the line length limit.
  - Otherwise they're each on their own line.
- Attribute set arguments must always start on a new line and they must not be mixed with identifier arguments.
  - If they have few attributes, the argument may be written on a single line
  - Otherwise each attribute must be on its own line with indentation, followed by a trailing comma.

**Examples**

```nix
#1
name: value: name ++ value

#2
name: value:
{
  "${name}-foo" = value;
}

#3
{ pkgs }: pkgs.hello

#4
args@{
  some,
  argument,
  default ? value,
  ...
}:
{
  # body
}

#5
{ pkgs }:
name: value: 
{
  # body
}

#6: These would be over the line length limit on a single line
aaaa:
bbbb:
cccc: 
dddd:
eeee:
null

#7: @ patterns can also come afterwards
{ pkgs }@args: pkgs.hello
```

**Alternatives**

- Have leading commas for parameters in attribute set arguments, like currently done in Nixpkgs.

> 

Reviewer 33: Little sad to see this go, but I think your points are valid in terms of normalization expression expansion.

Great work, having normalized formatting will remove a LOT of bike-shedding.

- Have the `then` on the start of the next line, directly followed by the if body:
  ```nix
  #1b
  if builtins.length matches != 0
    then { inherit path matches; }
  else if path == /.
    then [
      1
      2
    ]
  else go (dirOf path);

  #1c
  if builtins.length matches != 0
  then { inherit path matches; }
  else if path == /.
  then [
    1
    2
  ]
  else go (dirOf path);
  ```

### assert

- `assert`s must always start on their own line and the body also starts on its own line without any additional indentation.

```nix
# Good
assert foo;
[
  bar
  baz
]

# Bad
assert foo; [
  bar
  baz
]

# Good
{
  vendor ?
    assert false;
    null,
    
  vendor ? null,
}:
null

let
  # Good
  x =
    assert foo;
    bar;
   
  # Bad
  y = assert foo;
    bar;
in
x
```

**Alternatives**
- Treat it the [same as `with`](#with). The reasons not to do that:
  - `assert`'s stand on their own and could be removed without breaking anything. Comparatively, `with`'s can't be removed without breaking the code
  - `assert`'s are a bit like `if-then-else` statements, which are also spread out over multiple lines

### with

> - If the body is `{ ... }`, `[ ... ]`, `( ... )` or `'' ... ''`, then the body must start on that same line too.

Reviewer 17: But `++` binds more tightly than a `with` clause, so you're proposing putting the initial delimiter of a *subexpression* of the `with` body on the same line as `with`, which seems contrary to the spirit of this style guide. The current formatting nudges the reader toward understanding that the `with pkgs` extends over all of the concatenated lists and not only the first one.

- To allow having the `<body>` be on the same line as the `in`:
  ```
  let
    <name1> = <value1>;
    <name2> = <value2>;
    ...
  in <body>
  ```

  In particular when `<body>` is an identifier, list, attribute set and/or others.

  Problems with this alternative:
  - It leads to larger diffs when inserting something after the `in`
  - The formatting can change when `<body>` is updated
  - It's less consistent, since the formatting depends on the `<body>`

- The body could be indented by a level
  ```
  let
    <name1> = <value1>;
    <name2> = <value2>;
    ...
  in
    <body>
  ```
  
  Problems with this alternative:
  - Leads to indentation creeps
  - Inconsistent with other expressions that have a `<body>` that is "returned"
  - Favors a style where the body starts on the same line as the in for some values (e.g. attribute sets) to reduce an indentation level, see above.

### Attribute sets and lists

- Brackets and braces must always have a space (or line break) on the inside, like `[ `, ` ]`, `{ ` and ` }`.
  - Empty lists and attribute sets are written as `[ ]` and `{ }`, respectively.
- Lists and attribute sets with multiple items should be liberally expanded.

>   - They can only be on a single line if they fit on the line and contain few enough elements.

Reviewer 2: This seems to be a rule that would have to take the _input_ formatting into account. We're trying to really minimise such rules since they can lead to differing opinions. However we do already allow empty newlines from the input format to be persisted. And what you're suggesting feels like another good case where we can make an exception and allow it.

We'll give this a try and follow up later!

>   - They can only be on a single line if they fit on the line and contain few enough elements.

Author: After sleeping over the idea, an alternative to tracking the input would be to implement formatting directives. They have been discussed before and are left open as an option in the RFC. I'd honestly prefer that over adding another "impurity" to the parser.

Please provide some real-world examples of situations where you'd want the opposite of what Nixfmt does w.r.t. expansion, to better understand the tradeoffs between each of the approaches.

>   - They can only be on a single line if they fit on the line and contain few enough elements.

Reviewer 6: The most critical IMO is that it forces this declaration:

```Nix
    custom.packages.allowedUnfree = [
      "spotify"
    ];
```

into

```Nix
    custom.packages.allowedUnfree = [ "spotify" ];
```

while the list being multi-line was intentional to allow adding additional elements without diff noise.

>   - They can only be on a single line if they fit on the line and contain few enough elements.

Reviewer 17: Is this still preserved, and if so is it good enough for that case?
```nix
custom.packages.allowedUnfree = [
  "spotify"
  # ...
];
```

>   - They can only be on a single line if they fit on the line and contain few enough elements.

Author: Yes, as a work-around, adding a comment would force the multi-line layout. An empty line should work too btw (FWIW):

```nix
custom.packages.allowedUnfree = [

  "spotify"
];
```

While I make plenty of use of comments to force multiline layouts in the test suite, generally this isn't that great and we should try to find a better solution. As for the comment directives proposal, without giving too much thought to syntax for now it could look like this:

```nix
# @nixfmt:expand
custom.packages.allowedUnfree = [
  "spotify"
];
```

-----

However, I'd argue that a single-line `custom.packages.allowedUnfree = [ spotify ];` is not *that* bad. Having a single item with the expectation to add more in the future is pretty rare IME. Most of the time it stays at one item, or it already is more than one in the first place. So even when you run into the case where you only have one item and expect to add more in the future, as soon as the second item is added it will use the expanded layout anyways.

>   - They can only be on a single line if they fit on the line and contain few enough elements.

Reviewer 30: > Having a single item with the expectation to add more in the future is pretty rare IME

I would like to offer my own (anecdotal) experience as counter-point. People who learn nix/are not that experienced with it (a lot of whom use flakes for devshells for example) do this in my (again anecdotal) experience.

I myself do the same with my own nix configs, for purposes just like allowedUnfreePackages.

>   - They can only be on a single line if they fit on the line and contain few enough elements.

Reviewer 17: Advantages of an inline comment over a comment directive, in this particular case:
* No special syntax to remember
* Comment is literally a placeholder for more things, in exactly the place that the new things would go
* If `...` isn't evocative enough for you, you can make it say literally anything you want

Disadvantages:
* Less consistent with other cases where a comment directive might be more appropriate?

>   - They can only be on a single line if they fit on the line and contain few enough elements.

Reviewer 2: We discussed this today and agreed that it would be okay to preserve such multiline single-element lists. There's nothing in the RFC that disallows this, and we added a line that indicates that it would be allowed: https://github.com/NixOS/rfcs/pull/166/commits/463a858e57e4d68f4bfb1d38f20d8ac852ac4574

  - As described under [bindings](#bindings) below, nested attribute sets are always expanded.

**Examples**

```nix
[
  { }
  { foo = "bar"; }
  {
    foo = {
      bar = "baz";
    };
  }
  { foo.bar = "baz"; }
]

[
  [ 1 ]
  [
    2
    3
  ]
]

[
  [
    1
    2
    3
  ]
]

[
  {
    mySingletons = [
      [
        ({
          # stuff in there
        })
      ]
    ];
  
    mySingletons' = [
      [
        (function call)
      ]
    ];
  }
]
```

**Drawbacks**

- Singleton lists may use a lot of indentation

**Alternatives**

- Have a special compact form for singleton lists, to reduce the indentation level and remove two additional lines
  ```nix
  foo = [ {
    # content
  } ];
  ```

### Bindings

> Let bindings and attribute sets share the same syntax for their items, which is discussed here together.

Reviewer 2: This seems to have been caused by the line length limit: Everything on a single line would not fit under 100 chars.

But yeah, the default arguments really do use the same format as bindings, we also clarified this as part of a new commit: https://github.com/NixOS/rfcs/pull/166/commits/6b430d7f7efd04e454c54be215ef8c7195695f36

Bindings have the most special cases to accommodate for many common Nixpkgs idioms.
Generally, the following styles exist, which are used depending on the kind and size of the value:

> TODO: This entire section doesn't have any real specification and is rather unclear on details

Reviewer 2: Now resolved with https://github.com/NixOS/rfcs/pull/166/commits/6b430d7f7efd04e454c54be215ef8c7195695f36 :)

> ```nix

Reviewer 2: This seems to be specifically about chainable operators, which are specified in [this section](https://github.com/nix-rfc-101/rfcs/blob/master/rfcs/0166-nix-formatting.md#chainable-operators). Do you think this is sufficiently specified like that?

> ```nix

Reviewer 2: Now added @Atemu's examples to the Chainable operators section: https://github.com/NixOS/rfcs/pull/166/commits/3a91c378bee4f275629a2a347fe25e4ac1b5d6c5 :)

```nix
some.very.long.attr = callFunction
  arg1
  arg2
  arg3;
```

#### Bindings semicolon placement

The semicolon in bindings must always be placed on the same line as the expression it concludes.

**Examples**

```nix
{
  attr1 = bar;
  attr2 = function call {
    # stuff
  };
  attr3 =
    function call
      many
      arguments;
  attr4 =
    let
      foo = "bar";
    in
    some statement;
  attr5 =
    if foo then
      "bar"
    else
      "baz";
  attr6 =
    let
      foo = false;
    in
    if foo then "bar" else "baz";
  attr7 = function (
    if foo then
      "bar"
    else
      "baz"
  );
  attr8 =
    cond1
    || cond2
    ||
      some function call
      && cond3;
}
```

**Alternatives**

1. On a new line without indentation.
  - This clearly marks a separation between attributes, however it is wasteful of space.
  ```nix
  attr3 =
    function call
      many
      arguments
  ;
  attr3 =
    let
      foo = "bar";
    in
    some statements
  ;

>   ```

Reviewer 2: Just to clarify, you're suggesting that the formatter should _always_ add parenthesis for multi-line bindings? It sounds very controversial and non-standard, but I like the idea!

>   ```

Reviewer 6: Actually, thinking about it again, with a let binding with single line after `in` would probably be okay to have to terminator on that line because it statement is always the actual value of the binding (the let block just elaborates it).  
With other statements broken up into multiple lines such as ternaries, function calls or operator chainings, the order may change and the last line does not necessarily represent the true value of the binding.

I'm not sure about proposing it as the "chosen" style but I wanted to at least have it on the record as an alternative to think and reason about or perhaps take inspiration from. You can always choose to format it like this on your own and it'd still be valid under the new rules, even if the formatter doesn't automatically produce code like this.

>   ```

Reviewer 9: In my opinion bindings are easy to parse visually. We shouldn’t add visual noise to them. The semicolon should just be at the end of the last line of the binding. The semicolon is there to help the parser, not humans.

>   ```

Reviewer 9: If we waste a line for the `;` I don’t care very much, if we use parentheses or not. I will just note that editing a file with additional syntax requires slightly more typing.

>   ```

Reviewer 2: Now added this idea as an alternative: https://github.com/NixOS/rfcs/pull/166/commits/aaf1613f51abb97be149a5f429634a25277c2327

   These are the only syntax elements that may result in the semicolon being placed on a line with arbitrarily deep indentation.
   
   ```nix
   attr4 =
     if foo then
       "bar"
     else
       "baz"
   ;

   attr5 =
     let
       foo = false;
     in
     foo || bar;
       
   attr7 =
     cond1
     || cond2
     ||
       some function call
       && cond3
   ;
   ```

### inherit

The items must either be all on the same line, or all on a new line each (with indentation),
in which case the semicolon must be on its own line with indentation.

**Examples**

```nix
inherit foo bar baz;
inherit
  foo'
  bar'
  baz'
  ;
```

#### inherit from

For a fragment like this:
```
inherit (<source>) <attr1> ... <attrn>;
```

- If the entire fragment fits in the first line, it must be formatted as such.
- Otherwise if only `inherit (<source>)` fits into the first line, it must be formatted as such,
  with the same style as the normal `inherit` for the attributes.
- Otherwise the `(<source>)` must also be on its own line.

**Examples**

```nix
inherit (pkgs) ap1 ap2 ap3;
inherit (pkgs)
  app1
  app2
  # …
  app42

>   ;

Reviewer 24: Thouch a bit unintuitive, placing the semicolon in a separate line avoids unnecessary merge conflicts that happens from time to time inside the current Nixpkgs codebase.

>   ;

Reviewer 2: It's an interesting idea but we consider sorting items in such ways out of scope for the RFC, especially since it changes the AST.

- Having a commit that changes the formatting, can make git blame harder to use.
  - This can be [worked around in interfaces](#git-blames).

## Alternatives

- Keep the status quo of not having an official formatter.
  The danger is that this creates discord within the Nix community.
  The current friction and maintainer churn due to bad formatting may arguably be small, but not negligible.
- Specify a different format.
- Specify a format, but don't enforce it in CI, to allow manually tweaking the output if necessary.
- Apply the format incrementally, i.e. only changed files get formatted, to make a more graceful transition.
  However, such an approach would not reduce the amount of merge conflicts,
  and increase the workload for contributors during that time significantly.

## Unresolved questions

## Future work

- General style guidelines beyond AST reformatting
- Making widespread use of linters like Nixpkgs-hammering
- Enforcing the format to other official repositories

-----

## Terms and definitions:

- **Brackets:** `[]`
- **Braces:** `{}`
- **Parentheses**: `()`
- **Expressions:**
  All syntax nodes that would be a syntactically correct Nix program on their own.

> - **Terms:** The follwoing expressions are called terms 

Reviewer 34: ```suggestion
- **Terms:** The following expressions are called terms 
```

Reviewer 22: Fixed in 1e5914d2a214f5b3bf1e3a1279fd683cb7b18c2b

- Otherwise, the first argument not fitting onto the first line will start a new line with indentation, and all subsequent arguments will start on their own line as well.
  - All arguments that are not on the same line as the function must be indented by one level.

**Examples:**

```nix
# All arguments fit onto the first line
function arg1 arg2

# The line length limit is reached, so the remaining arguments need to be on their own lines
function arg1 arg2 arg3
  arg4
  arg5

# The last argument is a multiline expression, so it doesn't fit on the first line,
# but it can still start on the first line
function arg1 arg2 {
  more = "things";
}

# The second argument doesn't fit on the first line, but it's not the last argument,
# so it needs to start on a new line
function arg1 {
  more = "things";
} arg3

# In this case, the remaining arguments after the second one woulnd't fit onto the line anymore, therefore start all of them on a new line
function arg1
  {
    more = "things";
  }
  arg3
  many
  long
  args

# Same with multiple multiline arguments
function
  {
    a = 1;
    b = 2;
  }
  {
    c = 1;
    d = 2;
  }

> # Assume that the line lenght limit is here   ↓

Reviewer 34: ```suggestion
# Assume that the line length limit is here   ↓
```

Reviewer 22: Fixed in 1e5914d2a214f5b3bf1e3a1279fd683cb7b18c2b

  }:
  
  args@{
    some
  , argument
    # Single line comment
  , commentedArgument
  , # Comment on the value
    # multiline comment
    default ? value
  , ...
  }:
  # ...
  ```

  Problems with this alternative:
  - Moving items around with this style may require editing lines.
  - Inconsistent with the [expression expansion guidelines](#expansion-of-expressions), which disallows forms like `{ some`; `some` should start on a new line instead.
  - This does not work well with leading `@` bindings.
  - It's unclear whether comments belong to the next or the previous argument.
  - The leading comma style was a lesser-evil workaround for the lack of trailing commas in the Nix language. Now that the language has this feature, there is no reason to keep it that way anymore.

### Operators

From the [list of operators](https://nixos.org/manual/nix/stable/language/operators.html#operators), this section focuses on binary operators.
Function application and attribute selection are not treated as an "operator" in the sense of this section, see [function application](#function-application) instead.

#### Non-chainable operators

Operators with no associativity are non-chainable.

> Each invokation will always have exactly one left-hand side and one right-hand side.

Reviewer 34: ```suggestion
Each invocation will always have exactly one left-hand side and one right-hand side.
```

Reviewer 22: Fixed in 1e5914d2a214f5b3bf1e3a1279fd683cb7b18c2b

> ---

Reviewer 4: I've noticed that lists of attrsets look kinda bad for example 
https://github.com/NixOS/nixpkgs/blob/75f9d4f7cf99764b304d60fc6c8963ce565ed2f5/doc/default.nix#L10-L29
turns into 
https://github.com/piegamesde/nixpkgs/blob/e345d638168c066fae7e5bcfee414eae919712b1/doc/default.nix#L13-L79
```diff
     libsets = [
-      { name = "asserts"; description = "assertion functions"; }
-      { name = "attrsets"; description = "attribute set functions"; }
-      { name = "strings"; description = "string manipulation functions"; }
-      { name = "versions"; description = "version string functions"; }
-      { name = "trivial"; description = "miscellaneous functions"; }
-      { name = "fixedPoints"; baseName = "fixed-points"; description = "explicit recursion functions"; }
-      { name = "lists"; description = "list manipulation functions"; }
-      { name = "debug"; description = "debugging functions"; }
-      { name = "options"; description = "NixOS / nixpkgs option handling"; }
-      { name = "path"; description = "path functions"; }
-      { name = "filesystem"; description = "filesystem functions"; }
-      { name = "fileset"; description = "file set functions"; }
-      { name = "sources"; description = "source filtering functions"; }
-      { name = "cli"; description = "command-line serialization functions"; }
-      { name = "gvariant"; description = "GVariant formatted string serialization functions"; }
-      { name = "customisation"; description = "Functions to customise (derivation-related) functions, derivatons, or attribute sets"; }
-      { name = "meta"; description = "functions for derivation metadata"; }
-      { name = "derivations"; description = "miscellaneous derivation-specific functions"; }
+      {
+        name = "asserts";
+        description = "assertion functions";
+      }
+      {
+        name = "attrsets";
+        description = "attribute set functions";
+      }
+      {
+        name = "strings";
+        description = "string manipulation functions";
+      }
+      {
+        name = "versions";
+        description = "version string functions";
+      }
+      {
+        name = "trivial";
+        description = "miscellaneous functions";
+      }
+      {
+        name = "fixedPoints";
+        baseName = "fixed-points";
+        description = "explicit recursion functions";
+      }
+      {
+        name = "lists";
+        description = "list manipulation functions";
+      }
+      {
+        name = "debug";
+        description = "debugging functions";
+      }
+      {
+        name = "options";
+        description = "NixOS / nixpkgs option handling";
+      }
+      {
+        name = "path";
+        description = "path functions";
+      }
+      {
+        name = "filesystem";
+        description = "filesystem functions";
+      }
+      {
+        name = "fileset";
+        description = "file set functions";
+      }
+      {
+        name = "sources";
+        description = "source filtering functions";
+      }
+      {
+        name = "cli";
+        description = "command-line serialization functions";
+      }
+      {
+        name = "gvariant";
+        description = "GVariant formatted string serialization functions";
+      }
+      {
+        name = "customisation";
+        description = "Functions to customise (derivation-related) functions, derivatons, or attribute sets";
+      }
+      {
+        name = "meta";
+        description = "functions for derivation metadata";
+      }
+      {
+        name = "derivations";
+        description = "miscellaneous derivation-specific functions";
+      }
     ];
```
which uses a lot more vertical space and also looks kinda bad (to me).

I don't really have a solution for that tho.

maybe:
```nix
[
  { name = "asserts";
    description = "assertion functions"; }
  { name = "attrsets";
    description = "attribute set functions"; }
]
```
but i think that doesn't fit at all with the general style of the rfc.

Any ideas or thoughts on that?

Reviewer 21: While I'd agree that the re-format reduces readability, not sure how to make this a consistent rule.

Reviewer 6: This is a similar to [my comment about lists](https://github.com/NixOS/rfcs/pull/166#discussion_r1480439213).

I'd therefore also generalise this rule to mean that you MAY have any number of elements in a list or attrset as long as it fits the line limit but you SHOULD expand it when the addition of more elements in the future is likely.

For the formatter, it'd have to take previous formatting into account. For transitioning a non-compliant code-base, it'd be good to have a flag for a stricter mode which forces all MAYs to SHOULDs. Humans can then undo that for the specific edge cases where single line formatting is just better.

Reviewer 6: No matter how we ultimately go about it, I'd like for these downsides and trade-off discussion to be added to the RFC text.

Author: I don't see any way to make a rule which would improve the formatting of this code example without making other cases worse.

And given that boilerplate is rarely pretty in nature and should generally be avoided, I am okay with expanding the items here.

We already amended the RFC to allow implementations to preserve the input formatting in limited circumstances. I think that the advantages and downsides of our approach have been sufficiently discussed on a general level in the RFC text, and that there is no need to rehash them for every example somebody finds.

Reviewer 2: The current rules would allow persisting the original formatting, since it only says "few enough":

https://github.com/NixOS/rfcs/blob/aaf1613f51abb97be149a5f429634a25277c2327/rfcs/0166-nix-formatting.md?plain=1#L1403

So while the current implementation doesn't do that, we could easily change this in the future.

> ---

Reviewer 6: Editors plugins are up to the individual projects. I don't see them as part of the scope of this RFC.

The only thing that is explicitly part of the RFC is the reference formatter. You can simply run that executable using generic editor plugins.

  This allows the formatter to expand or compact multi-line expressions, while still allowing grouping of code.

  For example, formatting this code:
  ```nix
  [
    0 10

    (
      20 + 1
    )

    30
  ]
  ```

  turns into this:
  ```nix
  [
    0  # Line break added
    10

    (20 + 1) # Line breaks removed
             # Consecutive empty lines turned into a single empty line
    30
  ]
  ```

- Expressions of the same kind that can be treated as a sequence of expressions on the same level should be treated as such, even though they are technically parsed as a nested tree.
  - This applies to else-if chains, functions with multiple arguments, some operators, etc.
  - Example:
    ```nix
    # This is treated as a sequence of if-then-else chains, instead of indenting the second if as part of the else body
    if cond1 then
      foo
    else if cond2 then
      bar
    else
      baz
    ```

- Indentation should reflect the expression structure.
  Example:
  ```nix
  # Bad, the indentation misleads the user
  {
    foo = {

>     bar = if

Reviewer 6: This would violate these specifications of the bindings section:

> The semicolon in bindings must always be placed on the same line as the expression it concludes.

The semicolon must be behind the `20`.

> Within bindings, if the first and last line are not indented, the absorbed style is used, otherwise newline and indent

Your example uses absorbed style while the last line with an expression (`20;`) is indented.

What this rule together with the single indentation level growth rule effectively prevents is the placement of structural element like `else` or `in` on the top-level indent of the keys. See how your `else` is on the same indent as `bar`?

It's a bit unintuitive, so perhaps this "don't encroach on the keys indent" rule could be added as an explicit guideline.

>     bar = if

Reviewer 2: We now improved the clarity of the bindings rule and added your example with https://github.com/NixOS/rfcs/pull/166/commits/685ee9bf80abdfaceffe1e423f9c0a0609b82890 :)

>     bar = if

Reviewer 6: @infinisil I would like you to clarify: Is the effect that attribute names have an exclusive indent level intended?

If so, could that be added as an explicit rule? The rules of the RFC appear to indirectly cause this effect but it's not obvious that they do or whether that's the reason for their existence. This could make some of them feel less arbitrary.

    baz == null then 10
      else 20
    ;
  }; }

  # Good
  {
    foo = {
      bar =
        if baz == null then
          10
        else

>           20;

Reviewer 2: See https://github.com/NixOS/rfcs/pull/166#discussion_r1501819203 for justification, we think this is fine.

- There may also be a configurable _hard_ line length limit, which includes the leading indentation.
- String-like values such as strings, paths, comments, urls, etc. may go over the hard line length limit.

### Indentation

- Two spaces must be used for each indentation level.
  - This may be revisited should Nix get proper support for [using tabs for indentation](https://github.com/NixOS/nix/issues/7834) in the future.
- Vertical alignment must be ignored, both at the start of the line and within lines.
  - Examples:
    ```nix
    {
      # Bad, vertical alignment Within lines
      linux    = { execFormat = elf;     families = {              }; };
      netbsd   = { execFormat = elf;     families = { inherit bsd; }; };
      none     = { execFormat = unknown; families = {              }; };
      openbsd  = { execFormat = elf;     families = { inherit bsd; }; };

      # Bad, vertical alignment at the start of line
      optExecFormat =
        lib.optionalString (kernel.name == "netbsd" &&
                            gnuNetBSDDefaultExecFormat cpu != kernel.execFormat
                           )
                           kernel.execFormat.name;

>     }

Reviewer 6: For sections like these, it'd be great if you could make explicit exceptions to this rule. There's like a handful of places this kind of tabular style really has its benefits, most of them in the systems declarations.

>     }

Reviewer 35: Nix language imho plays the declarative config management role here. Other declarative config DSLs allow (and some even require) vertical "tabular" vertical alignment as in the `linux, netbsd, ...` case, examples being Puppet or OpenTofu (Terraform) where the latter is even conceptually quite close to Nix.

It is unclear to me whether this kind of alignment is "bad" or "discouraged" and/or whether the formatter should or should not try to "fix" it.

The statement says that formatter should in fact _ignore_ this kind of vertical alignment which to me means that after the code is formatted, vertical alignment should be _preserved as is_. On the other hand the example code says vertical alignment is "bad" and so it is (to me at least) ultimately unclear what the point is.

I personally like and use vertical alignment as in the first example because it (imho) _significantly_ increases human readability of the _declarative config_ code by allowing much faster and effective "eye scanning" of the code.

Note that I'm talking about the _config management_ role of the Nix language, not the _programming_ role - it is imho important to remember that Nix has multiple roles which may justify different formatting styles based on the context the code is used in.

If the statement (as I hope) really means that vertical alignment should be preserved by the formatter then I suggest removing the comment stating that the formatting is "bad".

If not then I would very much like to see some objective reasoning as to _why_ is it "bad", because to me, obviously, it is actually (very) good.

>     }

Author: > The statement says that formatter should in fact ignore this kind of vertical alignment which to me means that after the code is formatted, vertical alignment should be preserved as is.

I agree that the wording of the text needs to be adjusted, because this is certainly not its intent.

Let's put the discussion about whether we want vertically aligned code or not aside, and instead focus on on the *how*. Because simply saying "I want this specific code preserved" is not a rule that can be implemented. Everything needs to be algorithmically specified, and there is a strong requirement that doing this should not reduce the format quality for unaffected code. This is not a trivial problem.

We would probably also need a proper specification of what vertical alignment actually should look like, because if you look at Nix code you'll see various different styles being used. Since the point of this RFC is to unify the code style, this would have to be handled as well.

So even if we added an exception to allow for handling vertical alignment in some way (as we recently have done for selected other scenarios), this would still need someone to put the effort of actually implementing it.

>     }

Reviewer 35: > simply saying "I want this specific code preserved" is not a rule that can be implemented.

fair point

> a proper specification of what vertical alignment actually should look like, 

that's where I think the different contexts of where Nix lang is used comes in (_declarative config_ vs _programming_ vs possibly other use cases I can't think of right now) and I agree it is far from trivial, if not impossible to infer the context from the code alone without some kind of "tagging" or "metas" or somesuch. And that would be bad.

> Let's ... focus on on the how. 

I have a very naive perspective on this but one of the options I see is to, at least at first, _not try_ to solve the problem of _how should it look like_ and rather just trust the code owner to do the inline vertical alignment properly according to whatever the use case of the code is. The algorithmic part then would be rather simple one: formatter should ignore any whitespaces between symbols within a line, except at the beginning and the end of the line. This could even be an optional rule in the formatter that must be explicitly enabled by a flag or config with the intention to keep the rules clear for any upstream which chooses to adhere to them while still allowing private use cases to leverage the formatter without breaking their style.

I believe that this might not be so hard to implement if the formatter code is designed around individual rules.

>     }

Reviewer 2: We now updated this part with better justification and more examples: https://github.com/NixOS/rfcs/pull/166/commits/26b444b58db94baa2636d3b940a90768745e8692

>     }

Reviewer 6: One rather simple practical method of allowing free-form formatting for certain sections of code that just sprung to my mind would be to do so on file-level granularity. A certain comment text in the first line would simply disable the auto-formatter for the entire file. If the formatter would encounter such a comment, the entire file would be considered formatted correctly and remain untouched.

Sections like the systems example could then simply be placed into a dedicated file, formatted with vertical alignment and whatever other non-conformant arbitrary rules without upsetting future CI.

I imagine this could cover most of the exceptional cases where vertical alignment is absolutely necessary.

    ```
- Increasing indentation levels must not be "skipped": On subsequent lines, indentation can only increase by at most one level, but may decrease arbitrarily many levels.
  - Examples:
    ```nix
    buildInputs = [
        foo # <-- Bad, indentation increases by 2 levels

>       ] // lib.optionals cond [

Reviewer 36: Lists should be concatinated with `++`. Also I think the expression would require parentheses otherwise it would try to concatinate the function instead:
```
] ++ (lib.optionals cond [
```

>       ] // lib.optionals cond [

Reviewer 37: > Lists should be concatinated with `++`. Also I think the expression would require parentheses otherwise it would try to concatinate the function instead:
> 
> ```
> ] ++ (lib.optionals cond [
> ```

Right regarding `++`! Though the extra parenthesis are not required as is widely seen throughout Nixpkgs.

        bar
      ];

    attribute = { args }: let
        foo = "bar" # <-- Bad, indentation increases by 2 levels
      in
        foo;

    (callFunction {
        foo = "bar"; # <-- Bad, indentation increases by 2 levels
      }
      arg

>     )

Reviewer 2: Done with dec17975ac771f18ba7085c0af387481436a7aac :)

    # This is okay, indentation increases only one level per line
    let
      x = {
        a = foo
          bar

>           baz;

Reviewer 6: The alternatives section later explains that this is a trade-off between diff noise and vertical space. The new RFC style is already quite vertical, so I think it's valid to try and save on vertical space wherever possible.

For function calls, it's not as critical IMHO as functions don't typically change their amount of accepted parameters very often.

Previous discussion: https://github.com/NixOS/rfcs/pull/166#discussion_r1480448811

>           baz;

Reviewer 2: We now improved the semicolon placement alternatives section a bit, also adding @Atemu's suggestion from the linked thread: https://github.com/NixOS/rfcs/pull/166/commits/aaf1613f51abb97be149a5f429634a25277c2327

    - In this case, the inner term may be force-expanded.
    - This results in less indentation for many common Nix idioms.

**Examples:**

```nix
# All arguments fit onto the first line
function arg1 arg2

# The line length limit is reached, so the remaining arguments need to be on their own lines
function arg1 arg2 arg3
  arg4
  arg5

# The last argument is a multiline expression, so it doesn't fit on the first line,
# but it can still start on the first line
function arg1 arg2 {
  more = "things";
}

# The second argument doesn't fit on the first line, but it's not the last argument,
# so it needs to start on a new line
function arg1 {
  more = "things";
} arg3

# In this case, the remaining arguments after the second one woulnd't fit onto the line anymore, therefore start all of them on a new line
function arg1
  {
    more = "things";
  }
  arg3
  many
  long
  args

# Same with multiple multiline arguments
function
  {
    a = 1;
    b = 2;
  }
  {
    c = 1;
    d = 2;
  }

# Assume that the line length limit is here   ↓
# Good
concatMapString (s: "short string: ${s}") (
  attrsToList foo

> )

Reviewer 4: Other than that, until where I've come today in looking through formatted nixpkgs most code is really readable.
The few cases I found less readable weren't more readable before formatting, so it's probably just bad code. They were all cases of very large nested function calls.

There are quite a few cases of large increases of line count, but it always came with a readability improvement imo.

> )

Reviewer 2: We've adjusted the wording to be less strict about this and also allow your proposed alternative style with https://github.com/NixOS/rfcs/pull/166/commits/aa2141a9cad9a3e784c160913b042e3d27a15bbc :)

- Treat it the [same as `with`](#with). The reasons not to do that:
  - `assert`'s stand on their own and could be removed without breaking anything. Comparatively, `with`'s can't be removed without breaking the code
  - `assert`'s are a bit like `if-then-else` statements, which are also spread out over multiple lines

### with

- In any situation where a term would get absorbed, the term with a `with` prepended to it may get absorbed as well.
- Otherwise, the body of `with attrs;` must start on a new line without any additional indentation.

**Examples**

```nix
{
  # Good
  foo = with bar; [
    # multiline
    baz
  ];
  
  # Good
  foo =
    with foo;
    with bar;
    [
      # multiline
      baz
    ];
  
  # Good
  foo =
    with bar;
    baz foo {
      # multiline
      qux = 10;
    };

  # Good
  foo =
    with bar;
    if cond then
      foo
    else
      bar;
  
  # Bad
  foo = assert qux; with bar; [
    # multiline
    baz
  ];
  
  # Bad
  foo = with bar;
    [
      # multiline
      baz
    ];

  # Bad
  foo =
    with bar; [
      # multiline
      baz
    ];
    
  # Good
  [
    qux
    quux
  ]
  ++ (with pkgs; [
    baz
    blorp
  ]);
}
```

> ### let-in

Reviewer 26: I don't know if this was discussed already, but what about this example?

```
# let in the same line

foo = let
  A = "oof";
in
  A+A;

# let in the next line -- one more indentation
foo =
  let
    A = "oof";
  in
    A+A;
```

Is this expected?

Reviewer 6: See the bindings absorption rules for reference.

> Within bindings, if the first and last line are not indented, the absorbed style is used, otherwise newline and indent

The last line is always indented in a let binding.

  in
    <body>
  ```
  
  Problems with this alternative:
  - Leads to indentation creeps
  - Inconsistent with other expressions that have a `<body>` that is "returned"
  - Favors a style where the body starts on the same line as the in for some values (e.g. attribute sets) to reduce an indentation level, see above.

### Attribute sets and lists

- Brackets and braces must always have a space (or line break) on the inside, like `[ `, ` ]`, `{ ` and ` }`.
  - Empty lists and attribute sets are written as `[ ]` and `{ }`, respectively.
- Lists and attribute sets can only be on a single line if they fit on the line and contain few enough items.
- Lists and attribute sets with more items should be liberally expanded.
- As described under [bindings](#bindings) below, nested attribute sets are always expanded.

**Examples**

```nix
[
  { }
  { foo = "bar"; }
  {
    foo = {
      bar = "baz";
    };
  }
  { foo.bar = "baz"; }
]

[
  [ 1 ]
  [
    2
    3
  ]
]

[
  [
    1
    2
    3
  ]
]

[
  {
    mySingletons = [
      [
        ({
          # stuff in there
        })
      ]
    ];
  
    mySingletons' = [
      [
        (function call)
      ]
    ];
  }
]
```

**Drawbacks**

> - Singleton lists may use a lot of indentation

Reviewer 5: Didn't follow all conversations lately, but why don't we use an extra case for singleton lists?

Author: The wording here is a bit unfortunate (due to its surrounding context having changed over time), but it compares the current style to the alternative described a couple of lines down below. The current layout was chosen despite this drawback. Singleton lists that trivially fit onto one line as the one you show here are not affected by this decision btw. This is only about lists with a single but multi-line element.

Reviewer 38: In Nix's pretty-printer (unreleased as of yet) we print lists on a single line if they contain one item that isn't a list or set itself:

https://github.com/NixOS/nix/blob/eaa6c261899845bd4de80939831c6570c483e041/src/libexpr/print.cc#L364-L384

Reviewer 2: We don't think this needs to be further clarified for now.

  ```nix
  attr3 =
    function call
      many
      arguments
  ;
  attr3 =
    let
      foo = "bar";
    in
    some statements
  ;
  ```
2. On a new line with one indentation level.
  - Just as wasteful on space as (1), but a bit less clear about signaling the end of the binding.
  ```nix
  attr3 =
    function call
      many
      arguments
    ;
  ```
3. A mix of (1) and (2), where usually the semicolon is placed directly at the end of the binding.
   But with exceptions in which the semicolon is placed onto the following line instead in cases where the value is a multiline `if` expression or nested operator.
   These are the only syntax elements that may result in the semicolon being placed on a line with arbitrarily deep indentation.
   
   ```nix
   attr4 =
     if foo then
       "bar"
     else
       "baz"
   ;

   attr5 =
     let
       foo = false;
     in
     foo || bar;
       
   attr7 =
     cond1
     || cond2
     ||
       some function call
       && cond3
   ;
   ```

### inherit

The items must either be all on the same line, or all on a new line each (with indentation),

> in which case the semicolon must be on its own line with indentation.

Reviewer 6: With bindings, the trade-off was between wasted vertical space and diff noise in some expressions (if, sometimes functions) where the preference fell to vertical space savings.

The same trade-off applies here but the preference went the other way which is inconsistent. I think that should explicitly be called out and have a reason given.

I assume this was done because appending elements to an inherit happens quite frequently (i.e. `inherit (lib) foo bar ...;`).

Reviewer 2: Now added justification for this in https://github.com/NixOS/rfcs/pull/166/commits/f33b211ed60d0f12acad54bdfeecc0ca677a41fe :)
